{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# concrete.biopython, practical applications\n",
    "\n",
    "**concrete.biopython** is a **FHE** library based on python [**biopython**](https://biopython.org/) library. It implements the same objects and functions when they are compatible with fhe.\n",
    "\n",
    "In bioinformatics, the most basic object is the **sequence**. Sequences generally represent **DNA** and **RNA** strands, or **proteins** which are **amino acids** strands. Bioinformatics computations are most often processings of such sequences. This tutorial presents some basic processing algorithms on sequences, in **FHE**. Indeed, as data is often very sensitive in bioinformatics, using **FHE** guarantees end to end privacy of the data and of its processing.  \n",
    "\n",
    "Some applications are inspired from this python bioinformatics [tutorial](https://github.com/mrdbourke/coursera_bioinformatics_and_genetic_algorithm_experiment/blob/master/coursera-bioinformatics-content-and-genetic-algorithm-experiment.ipynb). We will first see how to make reusable code to easily generate circuits that can be used for different applications. Then we will see the applications themselves. We will display beforehand the difficulty level of each application.\n",
    "\n",
    "## Summary\n",
    "[I. Setting up FHE for practical applications](#section-i)  \n",
    "\n",
    "[II. Computing the distance between two sequences](#section-ii)  \n",
    "> [Hamming distance](#section-ii-1)  \n",
    "  [Levenshtein distance](#section-ii-2)\n",
    "  \n",
    "[III. Finding the origin of replication in a sequence](#section-iii)  \n",
    "\n",
    "[IV. Sorting sequences](#section-iv)  \n",
    "> [Comparing a list of sequences](#section-iv-1)  \n",
    "  [Ranking a list of sequences](#section-iv-2)  \n",
    "  [Sorting sequences of same length](#section-iv-3)  \n",
    "  \n",
    "[V. Aligning sequences](#section-v)  \n",
    "\n",
    "> [Local alignment with the Smith-Waterman algorithm](#section-v-1)  \n",
    "  [Global alignment with the Needleman-Wunsch algorithm](#section-v-2)  \n",
    "  [Going further: alignment software used in the industry](#section-v-3)  \n",
    "  [Bonus: improving the speed of the Levenshtein distance](#section-v-4)  \n",
    "\n",
    "<a id=\"section-i\"></a>\n",
    "## I. Setting up FHE for practical applications\n",
    "\n",
    "Before presenting the algorithms, we will set up some reusable code that will help us for each application later. Given a sequence processing function and a list biopython <span style=\"color:green\">**Seq**</span> objects, our reusable code should allow to:\n",
    " \n",
    "**i.** Generate automatically a circuit adapted to the length of our input and using a given sequence processing function.  \n",
    "\n",
    "**ii.** Run the circuit and measure performance on given input.\n",
    "\n",
    "First of all, we need to import **numpy** and **concrete.fhe**, as well as <span style=\"color:green\">**Seq**</span>  from **Bio.Seq**.\n",
    "Then we import <span style=\"color:#5CC8FF\">**SeqInterface**</span>, <span style=\"color:orange\">**FheSeq**</span>  and **Alphabets** from **concrete_biopython.FheSeq**, and also <span style=\"color:purple\">**BioCircuit**</span> from **concrete_biopython.BioCircuit** which will allow to create circuits easily. \n",
    "\n",
    "You can check out the quickstart tutorial presenting these objects in more details."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from concrete import fhe\n",
    "from Bio.Seq import Seq, MutableSeq\n",
    "\n",
    "import sys, os, time\n",
    "sys.path.append(os.path.dirname(os.getcwd()))\n",
    "\n",
    "from concrete_biopython.FheSeq import SeqInterface, Alphabets, FheSeq, FheMutableSeq\n",
    "from concrete_biopython.BioCircuit import BioCircuit"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then, we make a `compute_fhe_output` that creates a circuit for a given sequence processing function, and runs it on a list of <span style=\"color:green\">**Seq**</span> objects:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_fhe_output(seq_list, process_seq, description, seq_output=False, seq_interface=None,  res=False):\n",
    "    \"\"\"\n",
    "    Pack up the creation and running of a circuit,\n",
    "    and display the result along with the computation times\n",
    "    `seq_list`: a list of Seq or MutableSeq objects\n",
    "    `process_seq`: a function taking Seq or FheSeq objects that we want to compile for FHE\n",
    "    `description`: some description\n",
    "    `seq_output`: wether the output is a sequence or a regular integer array\n",
    "    `seq_interface`: a sequence interface for creating FheSeq with correct alphabet\n",
    "    `res`: wether to output the result or not\n",
    "    \"\"\"\n",
    "    # First print the description and a waiting message\n",
    "    print(description)\n",
    "    print(\"Computing ...\", end=\"\", flush=True)\n",
    "    print(\"\\r\", end=\"\")\n",
    "    \n",
    "    # prepare configuration\n",
    "    configuration=fhe.Configuration(\n",
    "        enable_unsafe_features=True,\n",
    "        use_insecure_key_cache=True,\n",
    "        insecure_key_cache_location=\".keys\",\n",
    "        dataflow_parallelize=False, # setting it to True may crash the kernel\n",
    "    )  \n",
    "    \n",
    "    # Create a BioCircuit wrapped circuit\n",
    "    circuit = BioCircuit(\n",
    "        function=process_seq,\n",
    "        len_seqs=[len(seq) for seq in seq_list],\n",
    "        seq_interface=seq_interface or SeqInterface(Alphabets.ASCII), # if None, use a complete Alphabet (slower)\n",
    "        show_timing=True,\n",
    "        configuration=configuration,\n",
    "        seq_output=seq_output\n",
    "    )\n",
    "\n",
    "    encrypted = circuit.encrypt(*seq_list)\n",
    "    fhe_res = circuit.run(encrypted)\n",
    "    output = circuit.decrypt(fhe_res)        \n",
    "\n",
    "    print('==> Result :', output, '\\n')\n",
    "\n",
    "    # In case we need the result\n",
    "    if res:\n",
    "        return output"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "That's it ! Before starting our first practical application, we can try out this code on test sequences and processing function, checking homomorphically wether the two first sequences of the list are identical."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FHE: Are the first two sequences equal ?:\n",
      "|  Compiling  : 0.49 s  |\n",
      "|  Encrypting  : 0.10 s  |\n",
      "|  Running  : 0.17 s  |\n",
      "|  Decrypting  : 0.00 s  |\n",
      "==> Result : 0 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Create a simple sequence processing function and run it in FHE on test sequences:\n",
    "\n",
    "# check wether two sequences are identical\n",
    "def process_sequence(seq1, seq2):\n",
    "    return seq1 == seq2\n",
    "\n",
    "seq1 = Seq(\"ACGTACGT\")\n",
    "seq2 = Seq(\"CGTACGTA\")\n",
    "compute_fhe_output([seq1, seq2], process_sequence, seq_output = False,\n",
    "                    seq_interface= SeqInterface(Alphabets.DNA_RNA),\n",
    "                    description='FHE: Are the first two sequences equal ?:')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"section-ii\"></a>\n",
    "## II. Computing the distance between two sequences\n",
    "\n",
    "As a first application, we will compute in FHE the **distance between two sequences**. A distance between two sequences **A** and **B** is a function that measures their similarity and obeys the following conditions:\n",
    "- If A = B, then *distance(A,B)=0*\n",
    "- If A â‰  B, then *distance(A,B)>0*\n",
    "- For all sequences A and B, *distance(A,B) = distance(B,A)*\n",
    "- For all sequences A,B, and C, *distance(A,C) <= distance(A,B) + distance(B,C)*\n",
    "\n",
    "Computing the distance between two biological sequences, for instance **DNA** strands, is often used in bioinformatics as a measure of similarity. We will see here the [**hamming distance**](https://en.wikipedia.org/wiki/Hamming_distance) and the [Levenshtein distance](https://en.wikipedia.org/wiki/Levenshtein_distance).\n",
    "\n",
    "<a id=\"section-ii-1\"></a>\n",
    "### Hamming distance \n",
    "\n",
    "( difficulty level: *easy* )\n",
    "\n",
    "The **hamming distance** between two sequences of **same length** is simply the number of positions where the two sequences are different. It is often used in bioinformatics to measure the similarity of biological sequences of same length such as **DNA** strands.  \n",
    "\n",
    "Let's define the `hamming` function which computes this distance between two arrays. The `hamming_adapter` allows to adapt inputs from either <span style=\"color:green\">**Seq**</span> or <span style=\"color:orange\">**FheSeq**</span> objects:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def hamming(arr1, arr2):\n",
    "    \"\"\"\n",
    "    Return the number of differences between arrays arr1 and arr2\n",
    "    \"\"\"\n",
    "    if arr1.size != arr2.size:\n",
    "        raise ValueError('arrays must have same size')\n",
    "        \n",
    "    return np.sum(arr1 != arr2)\n",
    "\n",
    "def hamming_adapter(seq1, seq2):\n",
    "    \"\"\"\n",
    "    Adapter for hamming distance of Seq objects\n",
    "    Convert Seq or FheSeq objects into integer arrays\n",
    "    \"\"\"\n",
    "    if isinstance(seq1, FheSeq) or isinstance(seq1, FheMutableSeq):\n",
    "        return hamming(seq1.to_array(),seq2.to_array())\n",
    "    elif isinstance(seq1, Seq) or isinstance(seq1, MutableSeq):\n",
    "        return hamming(np.array(list(str(seq1))), np.array(list(str(seq2))))\n",
    "    else:\n",
    "        return NotImplemented"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's measure the distance between two short strands of **DNA**, both in python and FHE:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python: Haming distance of S & T: 10\n",
      "FHE: Haming distance of S & T:\n",
      "|  Compiling  : 0.38 s  |\n",
      "|  Encrypting  : 0.10 s  |\n",
      "|  Running  : 0.35 s  |\n",
      "|  Decrypting  : 0.00 s  |\n",
      "==> Result : 10 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "S = Seq(\"AGGTTGGTAAAATGGTCCGTGGC\")\n",
    "T = Seq(\"ACCGTAATAAACGTGTCCGTTGG\")\n",
    "\n",
    "# run the circuit in python\n",
    "OUTPUT_PYTHON = hamming_adapter(S, T)\n",
    "print('Python: Haming distance of S & T:', OUTPUT_PYTHON)\n",
    "\n",
    "# run the circuit in FHE and compare the results\n",
    "compute_fhe_output([S, T], hamming_adapter, seq_output = False,\n",
    "                   seq_interface= SeqInterface(Alphabets.DNA_RNA),\n",
    "                   description='FHE: Haming distance of S & T:')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The distance is relatively fast to compute in FHE because it is a simple algorithm."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"section-ii-2\"></a>\n",
    "### Levenshtein distance\n",
    "\n",
    "( difficulty level: *medium* )\n",
    "\n",
    "The **Levenshtein distance** between two sequences is a bit more complex to compute. It measures **the number of simple operations** needed to transform a sequence into another one. These three simple operations are the **deletion** of an element, the **addition** of an element, and the **swapping** of two different elements. Check out the pseudo-code algorithm at: [Levenshtein distance](https://en.wikipedia.org/wiki/Levenshtein_distance).\n",
    "\n",
    "This distance is also sometimes used in bioinformatics to measure the similarity of biological sequences with different length.\n",
    "\n",
    "Again, we define the `levenshtein` function in a recursive fashion (the recursion is in python, but the compiled graph will not be recursive in FHE), as well as a `levenshtein_adapter` to adapt inputs. Note that in `levenshtein` function, we use an **if** condition. While this is forbidden in FHE on encrypted variables, in this case the condition is allowed because it concerns the size of the arrays which is not encrypted.  \n",
    "\n",
    "However, contrary to the original algorithm where the condition `(arr1[0] == arr2[0])` will stop the computation of the corresponding recursion branch if it is true, the FHE algorithm needs to run both branches of the computation. Indeed, the result of the condition being encrypted, both branches will run which will greatly slow the computation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python: Levenshtein distance of S & T: 2\n"
     ]
    }
   ],
   "source": [
    "def min_array(arr):\n",
    "    \"\"\"\n",
    "    Return the minimum value in an array using boolean mutliplication\n",
    "    \"\"\"\n",
    "    minval = arr[0]\n",
    "    for i in range(1, arr.size):\n",
    "        is_lt = arr[i]<minval\n",
    "        minval = is_lt*arr[i] + (1-is_lt)*minval\n",
    "    \n",
    "    return minval\n",
    "\n",
    "def levenshtein(arr1, arr2):\n",
    "    \"\"\"\n",
    "    Compute the Levenshtein distance between two array objects\n",
    "    \"\"\"\n",
    "    if arr2.size == 0:\n",
    "        return fhe.one()*arr1.size\n",
    "    elif arr1.size == 0:\n",
    "        return fhe.one()*arr2.size\n",
    "    else:\n",
    "        iseq0 = (arr1[0] == arr2[0])\n",
    "        lev1 = levenshtein(arr1[1:], arr2[1:])\n",
    "        lev2 = levenshtein(arr1[1:], arr2)\n",
    "        lev3 = levenshtein(arr1, arr2[1:])\n",
    "        minlev = min_array(fhe.array([lev1, lev2, lev3]))\n",
    "        return iseq0*lev1 + (1-iseq0)*(1+minlev)\n",
    "\n",
    "def levenshtein_adapter(seq1, seq2):\n",
    "    \"\"\"\n",
    "    Adapter for levenshtein distance of sequence objects\n",
    "    Convert Seq or FheSeq objects into integer arrays\n",
    "    \"\"\"\n",
    "    if isinstance(seq1, FheSeq) or isinstance(seq1, FheMutableSeq):\n",
    "        return levenshtein(seq1.to_array(),seq2.to_array())\n",
    "    elif isinstance(seq1, Seq) or isinstance(seq1, MutableSeq):\n",
    "        return levenshtein(np.array(list(str(seq1))), np.array(list(str(seq2))))\n",
    "    else:\n",
    "        return NotImplemented\n",
    "    \n",
    "S = Seq(\"ACGT\")\n",
    "T = Seq(\"CGTA\")\n",
    "\n",
    "OUTPUT_PYTHON = levenshtein_adapter(S, T)\n",
    "print('Python: Levenshtein distance of S & T:', OUTPUT_PYTHON)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's test this distance on very small DNA strands inputs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FHE: Levenshtein distance of S & T:\n",
      "|  Compiling  : 23.12 s  |\n",
      "|  Encrypting  : 0.07 s  |\n",
      "|  Running  : 4.73 s  |\n",
      "|  Decrypting  : 0.00 s  |\n",
      "==> Result : 2 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "compute_fhe_output([S, T], levenshtein_adapter, seq_output = False,\n",
    "                   seq_interface= SeqInterface(Alphabets.DNA_RNA),\n",
    "                   description='FHE: Levenshtein distance of S & T:')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The result is very long to compute ! This not due to FHE slowness, but to the algorithm itself which is very greedy. An improved, faster version of the algorithm is proposed in [Bonus: improving the speed of the Levenshtein distance](#section-v-4)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"section-iii\"></a>\n",
    "## III. Finding the origin of replication in a sequence\n",
    "\n",
    "( difficulty level: *medium* )\n",
    "\n",
    "As a second application, we will look in FHE for the [**origin of replication**](https://en.wikipedia.org/wiki/Origin_of_replication) in a DNA strand. The **origin of replication** (also called the **replication origin**) is a particular sequence in a genome at which **replication** is initiated, used when a cell divides. This area of the genome sometimes needs to be analyzed, but it needs to be found first.  \n",
    "\n",
    "In most area of DNA genomes, there is an equal proportion of nucleotides **'A'** (adenine), **'C'** (cytosine), **'G'** (guanine) and **'T'** (thymine). However, around the **origin of replication**, the proportion of guanine and thymine (G and T) **is greater** than the proportion of adenine and cytosine (A et C). This particularity is helpful to find the approximate localisation of the origin of replication.  \n",
    "\n",
    "In the function `minimum_skew` we take a look at the ratio between letter 'G' to letter 'C', to retrieve the location in the strand where this ratio is minimal. To that extent, we first create a lookup table to map letters 'G' to a 1 and letters 'C' to a -1, the other letters 'A' and 'T' being mapped to 0. Then, we sum up the values throughout the sequences, and keep the first index of the minimum value of this cumulated sum. We also return the whole cumulated sum array for plotting.\n",
    "\n",
    "We need to use the <span style=\"color:#5CC8FF\">**SeqInterface**</span> class in this case. This class is used internally within the <span style=\"color:purple\">**BioCircuit**</span> class, but we need to explicitly use it here in order to convert letters to integers, and to know the value of the maximum integer that can be encoded, which is linked to the length of the alphabet used:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def minimum_skew(seq):\n",
    "    \"\"\"\n",
    "    Return the index of the minimum skew in a sequence\n",
    "    Note: Only the first index of minimum skew is returned\n",
    "    \"\"\" \n",
    "    # First, create a lookup table to map G to 1 and C to -1\n",
    "    if isinstance(seq, FheSeq) or isinstance(seq, FheMutableSeq):\n",
    "        # get the seq_interface of the seq object\n",
    "        seq_interface = seq.get_seq_interface()        \n",
    "        CG_integers = seq_interface.to_integers(Seq('CG'))\n",
    "        letter_table = [0]*(seq_interface.max_integer()+1)\n",
    "        letter_table[CG_integers[0]]=-1\n",
    "        letter_table[CG_integers[1]]=1\n",
    "        letter_table = fhe.LookupTable(letter_table)\n",
    "        skew_counts = fhe.zeros(len(seq))\n",
    "    elif isinstance(seq, Seq) or isinstance(seq, MutableSeq):\n",
    "        letter_table = {'A':0, 'C':-1, 'G':1, 'T':0}\n",
    "        skew_counts = np.zeros(len(seq)).astype('int')\n",
    "    else:\n",
    "        return NotImplemented\n",
    "    \n",
    "    # Then compute the cumulated sum and keep the first index of the minimum\n",
    "    skew_counts[0] = letter_table[seq[0]]\n",
    "    minimum = skew_counts[0]\n",
    "    min_index = 0\n",
    "    for i in range(1,len(seq)):\n",
    "        # Compute cumulated sum\n",
    "        skew_counts[i] = skew_counts[i-1] + letter_table[seq[i]]\n",
    "        # and keep minimum\n",
    "        ismin = (skew_counts[i]<minimum)\n",
    "        minimum = min_array(fhe.array([skew_counts[i], minimum]))\n",
    "        min_index = i*ismin + min_index*(1-ismin)\n",
    "        \n",
    "    return min_index, skew_counts\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As mentioned, the function will only output the first index where the minimum value is reached. This is a limitation of FHE. It is indeed not possible to retrieve an arbitrary number of locations in FHE, for instance to return a list with all indices where the minimum is reached. We could however return an array of the size of the sequence containing ones where the minimum is reached, or, with very extensive computations, containing the indices of the minimum followed by zeros.  \n",
    "\n",
    "Another limitation here is the size of the sequence. If the size becomes too long, the values in `skew_counts` may become larger than the maximum capacity of FHE integers that have limited bitwidth. But this could be overcome with counting trick such as splitting the hundreds or thousands in a separate counter variable. \n",
    "\n",
    "Let's test our function on a clear sequence, and plot the cumulated sum of 'C' letters minus 'G' letters. Then we test it with encrypted data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python: The minimum skew is at index: 18\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjoAAAGwCAYAAACgi8/jAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/bCgiHAAAACXBIWXMAAA9hAAAPYQGoP6dpAABQHElEQVR4nO3deXhU9d0+/vvMTGayTzLZQxLIyiKLyBLWBJVH1KfWrdW29udSaouirUXbShcRH1tsa31al8e2VqTWqtBaaWtbvi6VJOwCgiKQjUBCCNmTSSbJrOf3R3ImCQTIJGfmzDlzv66Lq51kcuadcWDu+SzvjyCKoggiIiIiDdIpXQARERGRvzDoEBERkWYx6BAREZFmMegQERGRZjHoEBERkWYx6BAREZFmMegQERGRZhmULiCQPB4Pzpw5g5iYGAiCoHQ5RERENAqiKKKrqwvp6enQ6XwbowmpoHPmzBlkZmYqXQYRERGNQV1dHTIyMnz6mZAKOjExMQD6n6jY2FiFqyEiIqLRsFqtyMzM9L6P+yKkgo40XRUbG8ugQ0REpDJjWXbCxchERESkWQw6REREpFkMOkRERKRZDDpERESkWQw6REREpFkMOkRERKRZDDpERESkWQw6REREpFkMOkRERKRZDDpERESkWaoJOi+++CJmzpzpPb5h4cKF+Pe//610WURERBTEVBN0MjIy8NRTT+HAgQPYv38/rrrqKtx444347LPPlC6NiIiIgpQgiqKodBFjZbFY8Itf/AIrV64c1f2tVivMZjM6Ozt5qCeFFFEU4XSLMBpU89mGiMhrPO/fqvxXz+12480334TNZsPChQsveD+73Q6r1TrsD1Eo2rjzJCb/+N8orWhWuhQiooBSVdD59NNPER0dDZPJhFWrVuHtt9/GtGnTLnj/DRs2wGw2e/9kZmYGsFqi4PHmvlqIIrBlf53SpRARBZSqgs7kyZNx6NAh7N27F/fddx/uuusuHD169IL3X7t2LTo7O71/6ur4jzyFnjMdvahs6gYAlFW2wO1R7Ww1EZHPDEoX4Auj0Yi8vDwAwJw5c/DRRx/h17/+NX7729+OeH+TyQSTyRTIEomCztDpqs5eJz453YHZWfEKVkREFDiqGtE5l8fjgd1uV7oMoqBWMhB0dMLw20REoUA1QWft2rUoLS3FyZMn8emnn2Lt2rXYvn077rjjDqVLIwpaLrcHO6paAAC3z+tfo8YFyUQUSlQzddXU1IQ777wTDQ0NMJvNmDlzJv7f//t/+K//+i+lSyMKWofqOtDV50JcZBhWX5mHN/bV4VBdBzp6HIiLNCpdHhGR36km6Lz88stKl0CkOtI01ZK8RGTERyI/ORqVTd3YUdWCz81MV7g6IiL/U83UFRH5TpqmKi5IGva/nL4iolDBoEOkUW02Bz6p7wQAFA0EHOl/SyqaoeKm6EREo8agQ6RRZZXNEEVgSmoMUmLDAQDzsy0ID9Oh0WpHRWO3whUSEfkfgw6RRknrc4onJ3m/Fh6mx4KchIHvNylSFxFRIDHoEGmQxyOitKJ/W3lxftKw7xXlS+t0WgJeFxFRoDHoEGnQsbNWtHTbEWnUY86k4V2QpRGefTVt6HG4lCiPiChgGHSINEiatlqYkwCTQT/sezmJUZgQFwGH24M9J1qVKI+IKGAYdIg0qHSE9TkSQRC8X+f0FRFpHYMOkcZ0213Yf7IdwOB6nHNJX+e5V0SkdQw6RBqzu7oVLo+IiQmRmJQYNeJ9FuUlwKATUNNiQ21rT4ArJCIKHAYdIo2Rto1LXZBHEhsehism9i9SLqnkqA4RaReDDpGGiKLonY660LSVhMdBEFEoYNAh0pCTrT2oa+tFmF7AwtyEi95XCjq7qlrgcHkCUR4RUcAx6BBpSEl5/7TV3IkWRJkMF73vtLRYJEQZYXO4ceBUeyDKIyIKOAYdIg0prRzohjzCtvJz6XSC95DPUq7TISKNYtAh0gi7y43d1f0NAC+1PkdSVJAIACgpZ9AhIm1i0CHSiP0n29HrdCMpxoSpaTGj+pmlA4HoaIMVTV19/iyPiEgRDDpEGuE9rbwgCYIgjOpnEqNNmDHBDAAoY5dkItIgBh0ijZC2iRddpH/OSKTpK67TISItYtAh0oCznX04frYLggAszUv06WeLC5IB9Aclt0f0R3lERIph0CHSAGk0Z2ZGHOKjjD797OysOESbDGjvceJIfac/yiMiUgyDDpEGSMc4XOzYhwsJ0+uwOK+/uSC7JBOR1jDoEKmc2yNih9Q/p8C3aSuJtK6Hp5kTkdYw6BCp3OHTHejsdSI23IBZGXFjuobUd+fjuv5rERFpBYMOkcpJzf6W5ifBoB/bX+lMSyRyk6Lg9ojYVcVt5kSkHQw6RConbQsvGuO0lYTHQRCRFjHoEKlYR48Dh+s6APjeP+dc0kLmkvJmiCK3mRORNjDoEKnYjqoWeESgICUaaeaIcV2rMDsBRoMOZzr7UN3cLVOFRETKYtAhUjFpfc5YtpWfK8KoR2G2BQCwnYd8EpFGMOgQqZQoit71NFJ34/Eq5jZzItIYBh0ilSpv7EKj1Y7wMB3mToqX5ZpS0NlX04Y+p1uWaxIRKYlBh0ilpGmrhTkJCA/Ty3LNvORopJvDYXd5sOdEqyzXJCJSEoMOkUoNbisf//ociSAIg9vMK9hPh4jUj0GHSIV6HC58VNMOQJ6FyEMNrtNpkvW6RERKYNAhUqE9J1rhcHuQER+B7MQoWa+9KC8Rep2A6mYbTrf3yHptIqJAY9AhUqGh28oFQZD12uaIMMzOjAPA6SsiUj8GHSIVkrZ/yz1tJeH0FRFpBYMOkcqcarXhZGsPDDoBC3MT/PIY0oLkXVWtcLo9fnkMIqJAYNAhUpnSgdGcORPjERMe5pfHmDHBDEuUEV12Fz6u7fDLYxARBQKDDpHKlAysm5FzW/m5dDoBS/L6T0MvZZdkIlIxBh0iFXG4PNhd3R90/LU+R8LjIIhICxh0iFTkwKl22BxuJEYbMS0t1q+PtbSgf0Tn0/pOtHTb/fpYRET+wqBDpCLS6EpRfhJ0Onm3lZ8rOSbcG6Z2VHKbORGpE4MOkYp4t5VP9u+0lUR6HE5fEZFaMegQqUSTtQ/HGqwQBHgXCvtbUX5/0CmrbIbHIwbkMYmI5MSgQ6QSpQPTRzMmmJEQbQrIY86ZGI8oox4t3Q4cbbAG5DGJiOTEoEOkEqVD1ucEitGgw8Lc/tEjTl8RkRox6BCpgNsjoqwysOtzJFynQ0RqxqBDpAJH6jvR3uNEjMmAywcO3AyU4oERpIOn2tHV5wzoYxMRjReDDpEKSKMpi/MSEaYP7F/brIRIZCdGweURsau6NaCPTUQ0Xgw6RCoQ6G3l52KXZCJSKwYdoiDX2ePEx7XtAPx7vtXFFBUMnnslitxmTkTqwaBDFOR2VrfAIwJ5ydGYEBehSA0LchJg1Otwur0XJ1psitRARDQWDDpEQU6JbeXnijQaMC87flg9RERqwKBDFMREUVR8fY6E63SISI0YdIiCWFVTNxo6+2Ay6FCYbVG0Fml90J4TrehzuhWthYhotBh0iIKYNHpSmJOA8DC9orVMTolBSqwJfU4PPjrZpmgtRESjpZqgs2HDBsybNw8xMTFITk7GTTfdhPLycqXLIvIr77SVQruthhIEwVsH1+kQkVqoJuiUlJRg9erV2LNnD9577z04nU5cc801sNm4A4S0qdfhxt6a/pGT4oLAnFZ+KUVcp0NEF9BtdwVl+wmD0gWM1rZt24bd3rRpE5KTk3HgwAEUFRWN+DN2ux12u91722rl6ctjUdfWg7tf2Yf2Hnna/0eE6fHMbbNQmJMgy/XksGbzIWwPsjdvl9sDh8uDCXERyE2KVrocAMCSvEToBKCisRtX/M97slxzQY4FL3zlCgiCIMv1iEgZD285hE9Od+LJm6bj6qkpSpfjpZqgc67Ozk4AgMVy4QWaGzZswPr16wNVkmb9+cBpVDfLO3L26p5TQRN0TrbY8NeP65Uu44JumJUeNCEgLtKIZZOT8Z/jTWizOWS55r8+PYvyxi5MSY2V5XpEFHhOtwc7q1rRbXchMdqkdDnDqDLoeDwePPTQQ1i8eDGmT59+wfutXbsWa9as8d62Wq3IzMwMRImaIk1TfP/aKVg+NXlc1ypv7MIDr3+MHZUtcHtE6HXKv4FLv9+cifF46pYZClczXJheh4kJkUqXMczv/r85ONlqgxwj1Ov+/hl2VbeipLyZQYdIxT6u7UC33QVLlBEzJpiVLmcYVQad1atX48iRI9ixY8dF72cymWAyBVeyVJt2mwOfnO4AANw8ewJSzeHjul52YhRiww3o7HXi8OkOXJEVL0OV4yMtrF0+NQX5KTEKVxP8DHod8pLleZ7+a1oKdlW3orSyGd8szpXlmkQUeCUVTQAGpreD4APsUKpZjCx54IEH8M477+DDDz9ERkaG0uVoXllVC0Sxf2vxeEMO0P8muSS/f2FtSbnya2LsLrf3RO6iIFnwG0qkxc0f1bSjx+FSuBoiGqvSihYAwbFD9FyqCTqiKOKBBx7A22+/jf/85z/Izs5WuqSQUOqHrrzeLcqVygedAyfb0et0IynGhGlpnDoJtJzEKGTER8Dh9mDPiValyyGiMWjptuPT+v51s0uD8AOjaoLO6tWr8dprr+H1119HTEwMzp49i7Nnz6K3t1fp0jRLFMXBoCNjSpc+xR+u60BHjzwLWseqZMg5UsGy4DeUDO3NEwwjfETkux2V/aM509JikRwz/pF/uakm6Lz44ovo7OzEsmXLkJaW5v2zefNmpUvTrONnu9DUZUdEmB5zJ8m3libNHIGClGh4RGBHVYts1x0Lb9AJwk8hoaLIO8Kn7GuBiMYmWM7juxDVLEYOxiZEWie9eBfmJsBkkPf4geKCJFQ0dqOkvBmfm5ku67VHq9Hah+NnuyAIwFIFTwYPdYtyE2DQCahpsaG2tQdZQbbLjIguzOMZHPkvCtJ/R1UzokOBN/jilX+0o2jIOh2lQqwU5GZOMMMSZVSkBgJiwsNwxcT+EcOSIFi3RUSjd7TBilabA1FGPeZMVH4X7UgYdGhENrvLe3Bj8eTx9c4ZybxJFoSH6dBotaO8sUv264+GP9Yf0dhwnQ6ROg2O/CfCaAjOSBGcVZHidle3wukWkWmJwCQ/TCWEh+mxYKAzshJvbm6PiLKBNSFFDDqKk4LO7uoWOFwehashotEK9vU5AIMOXYC09bu4wH+7kZTcZv7J6Q509joRE27A5ZlxAX98Gm5aWiwSo42wOdw4cKpd6XKIaBS6+pw4OPD3tThI1+cADDp0Ad6UXiD/tJWkWMFmcdLvtzQ/EQY9/xooTacTvAsZeTI6kTrsqm6FyyMiOzEqqDcR8F94Os/JFhtOtfbAoBOwMNd/B29mK9gsLth3CYQi7wJ1Bh0iVSjx44YVOTHo0HmkqaS5k+IRbfJfBwKlmsV19DhwqK4DANfnBJOl+YkQhP5dHE1dfUqXQ0QXIYqi99/tYF6fAzDo0Ai8ox0BCAHSYwRyumJHVQs8IpCfHI30uIiAPS5dXEK0CdPT+089Lqtg80CiYHaixYb6jl4Y9TrvxpJgxaBDwww95DIQ266lZnEnW3twqtXm98cDuK08mBUrEHyJyHfSaM687HhEGoO79zCDDg1z4GQ7ehxuJEabMDXV/4dcDm0WF4i1GaIoDjn2gUEn2Ej/Tcoqm+H2sBs6UbAaujM32DHo0DBSZ9qigkTodIE55HLwU7z/pysqGrvRaLUjPEyH+dkWvz8e+WZ2VhxiTAa09zhxZOA0ZCIKLn1Ot3cDiRo+MDLo0DDexWUBfPEGsllcSUUTAGBBTgLCw+Q9v4vGL0yvw+K8/h0cnL4iCk4fnWxDn9ODlFgTJqfEKF3OJTHokJdSh1wGsllc6cCoEbeVBy9uMycKbtIH4qJ8/zWUlRODDnmVKnTIZaCaxfU4XNhXI53fxaATrIoK+kd0Pq7r715NRMFFDcc+DMWgQ16lCp79FIht5ntOtMLh9mBCXARyEqP89jg0PhnxkchNioLbI2JXFbeZEwWTMx29qGzqhk4AluQFd6NACYMOAZAOuVRuFb3ULO5YgxVNVv80i5OmrYonq2O4NZRJR49wnQ5RcJFG/mdlxiEuMnAj/+PBoEMA+g+57OhR7pDLoc3ipJEluZXw2AfVkKavSiqaIYrcZk4ULNS0rVzCoEMABkc7luQpd8hlsR8Xoda29qCmxQaDTsCivODu4kn9u+JMBh0aOvtQ1dStdDlEBMDl9qBMwSUOY8WgQwAGt10rmdKlhW3+aBYn9Qe6YmI8YsPDZL02yS88TI/CgbbynL4iCg6HT3egq88Fc0QYZmXEKV3OqDHoEDp7nEFxyOXlmf5rFsdjH9RHOhGZQYcoOEjbypfkJ0IfoIaycmDQoaA55NJfzeIcLo939w6DjnosGxjh21vThl6HW+FqiKhEpR8YGXQooKeVX4o/tpkfONUOm8ONhCgjpqX5//wukkduUjTSzeFwuDzYW9OqdDlEIa3N5sAnAyPtDDqkKkMPuQyGF6+3WVxtOzp75GkWV1o5GOQCdX4XjZ8gCN51W5y+IlJWWWUzRBGYkhqDlNhwpcvxCYNOiKto7MZZax9MhuA45FJqFucRgZ3V8mwz97YrL1BHcysaFIiO2UR0ad4+ZEHwgdhXDDohTpq2CqZDLqVmcXJsM2/q6sPRBiuAwJ7fRfJYlNe/6PFEsw11bT1Kl0MUkkRRHDYyrjYMOiEumKatJEOnK8bbLK5s4FPIjAlmJEabxl0bBZY5IgxXZMUBGJyCJKLAOtbQheYuOyLC9Jg7KV7pcnzGoBPChh5yGUwpvTDbIluzuMFPIZy2Uitp+oqnmRMpQ/pAvDA3ASZDcIz8+4JBJ4TtPdHmPeQyNyl4DrmUq1mc2yMO6Z+TLEttFHjSCN/OqlY43R6FqyEKPcHQUHY8GHRCWMmQbeXBdsilHM3ijtR3or3HiWiTAbMHpj9Ifaanm2GJMqLb7sLHtR1Kl0MUUrrtLhw41Q4guEb+fcGgE8KCuVuwHM3ipN9vcV4CwhQ6v4vGT6cTsNQbfJsUroYotOyuboXTLSLLEolJCZFKlzMm/Nc/RNW19eBEiw36ID3kcmizuD1jbBZXEkSNEGl8uM2cSBlDPxAH28j/aDHohCjpDWNOVnAecjm0WdxYFqF29jrxsXR+F7eVq97SgcXkR+qtaOm2K1wNUejQwgdGBp0Q5d1WPjl4X7zF4zgOYldVC9weEblJUci0qHO4lQYlx4TjsvT+4zvKuM2cKCBOtthQ29aDML2AhbnBN/I/Wgw6Icjh8mB3df90UDCPdoynWZyam1vRyKT/llKHViLyL+/I/8R4RJsMClczdgw6IehgbTu67S4kRBm9n5KDUWz42JrFiaLoPfYhGBda09gUFwxOZXo842skSUSXVqKR9hwMOiFIWvOyND8x6A+59C5CLR990Klq6saZzj4YDToUZqt3uJWGuyIrHlFGPVptDu+xHkTkH3aXe3DkX+UNVxl0QpAa1udIpBp3VY++WZz0+xVmWxBhVF8XTxqZ0aDDorzx91ciokvbf7IdvU43kmJMmJYWvCP/o8GgE2Kau+z47Ix6Drkc2izu4EDTqksJxvO7SB5F41igTkSjJ438F+Wrd1u5hEEnxEg7VqZPiFXFIZdDm8WNZp1On9PtPb+LQUd7igfC+cFT7bD2ORWuhki7BreVq3vaCmDQCTlqHO3wZZv5nhOtsLs8SDeHIy852t+lUYBlJUQiJzEKLo+IXVVjayRJRBfXaO3D8bNdEAR1jPxfCoNOCPF4RJRV9m/NDeZt5eeS/qKNplmctPU4GM/vInl4t5mznw6RX0gfKmdO6F86oHYMOiHkyJlOtNkciDYZcMXEeKXLGbWkGNOom8Wp/ZRdujTvCF95M0SR28yJ5KbGkf+LYdAJIdLiskW56jvksqjg0tvMT7f3oLpZOr9L/fPKNLLCHAuMeh3qO3pxosWmdDlEmuL2iNhROTgyrgXqerejcVHTtvJzSZ8syipbLtgsTpq2mp0ZB3NE8J3fRfKINBowP9sCwLf+SkR0aYdPd6Cz14mYcAMuz4xTuhxZMOiECGufEwdrOwCoa32OZGizOGl7/LmkaSutfAqhC5N2gnCdDpG8hjaUNahs5P9CtPFb0CVJh1zmqPSQy6HN4kZ6c3O6Pd5dOFqZV6YLk1rS7znRij6nW+FqiLSjZEj/HK1g0AkRWlhcVnyRdTof13agy+6CJcqIGRPMgS6NAqwgJRqpseHoc3q8fZOIaHw6ehw4XNcBQFsj4ww6IUAUxWHbrtVKCjoHa89vFicNty7JC/7zu2j8BEEYnL5il2QiWeyoaoFHBPKTo5EeF6F0ObJh0AkB1c3dqO/ohdGgwwIVH3KZablwszgtjFiRb6TpKx4HQSQPabRca/+OMuiEgJKB0RwtHHI50llHLd12fFrfCQBYqoF25TQ6S/ISoROAyqZunOnoVbocIlUTRdG7/lHNI/8jGXfQsVqt2Lp1K44dOyZHPeQHWhrtkH6H0orBZnFSz4dpabFIjglXrDYKLHNkmHf7K6eviManvLELjVY7wsN03vYNWuFz0Lntttvw/PPPAwB6e3sxd+5c3HbbbZg5cybeeust2Quk8elzurH3RP80jxZS+tBmcdXN/c3iBg+fU//vR77hcRBE8pA+LBRmJyA8TN0j/+fyOeiUlpZi6dKlAIC3334boiiio6MDzz77LJ588knZC6Tx2VvTBrvLgzRzOPI1cMjl0GZxpRXNA+d3aWfEinwztJGky+1RuBoi9dLSyP+5fA46nZ2dsFj632i2bduGW2+9FZGRkfjv//5vVFZWyl4gjc/QxWVaOeRy6GnmRxusaOl2IMqoxxwVnd9F8piZEYe4yDB09blwaGBbLBH5psfhwkc17QDU2Tn/UnwOOpmZmdi9ezdsNhu2bduGa665BgDQ3t6O8HCujwg2WlxcJv0ue2ta8e5nZwEAC3MTYTRwbX2o0esELMnjNnOi8dhzohUOtwcT4iKQkxildDmy8/md4aGHHsIdd9yBjIwMpKenY9myZQD6p7RmzJghd300DvUdvahq6oZeJ2Cxhg65HNosbuPOkwC0+SmERqd4hJ14RDR63pH/ydoZ+R/K56Bz//33Y8+ePdi4cSN27NgBna7/Ejk5OfjJT34ie4FDlZaW4oYbbkB6ejoEQcDWrVv9+nhqJ33CvVxjh1wObRbXbXcBAIo11K6cfCON8H1S34k2m0PhaojUp1Q6rVyj/476HHSeeOIJTJ06FTfffDOiowcXt1511VV4//33ZS3uXDabDbNmzcILL7zg18fRCq02fwIGm8UBQHZiFLIS1Hd+F8kjJTYcU1JjIIrwLkwnotGpbe1BTYsNBp2ARXnqbSh7MQZff2D9+vVYtWoVIiOHv7H09PRg/fr1eOyxx2Qr7lzXXXcdrrvuulHf3263w263e29brSOfeh1MXt9bi20D607Ga1+NdraVn0tqFucRgaJ87UzL0dgUFyTh+NkuPP1uOd46WK90OcOYI8Lw+A3TkBBtUroU0og2mwOP//0zdPQ6L33nS2jp6n+PvCIrHrHh2hn5H8rnoCOK4ohzeIcPH/buxgoWGzZswPr165UuY9RsdhfW/f0InG5RtmumxoZr8pBLc2QYFuUmYkdVC1ZMT1W6HFLYNZel4LelJ1DX1ou6tuDrklyQHI0Hr85XugzSiDf21eLvh8/Ies1rLkuR9XrBZNRBJz4+HoIgQBAEFBQUDAs7brcb3d3dWLVqlV+KHKu1a9dizZo13ttWqxWZmZkKVnRxu6tb4XSLSDeH45EVk2W55pyJ8dBr9JDLX3/pcpxs7eG2csKciRa8fm8hznb2KV3KMAdr2/HanlqUVjYz6JBspGUJdxRmyfLvX5TJgCsnJ1/6jio16qDzq1/9CqIo4mtf+xrWr18Ps3lwlMBoNGLSpElYuHChX4ocK5PJBJNJPcPF0q6Rq6em4JYrMhSuJvglRJs4HUBei3KDbwpz3iQLXttTi4O1HbD2OTU7NUCBY+1z4kBtf8+bVcW5yLRwfeKljDro3HXXXQCA7OxsLFq0CGFh/AsrNy32vCEKZZmWSOQkReFEsw27qlpw7fQ0pUsildtV1Qq3R0ROYhRDzij5vEanuLgYHo8HFRUVaGpqgsczvO16UVGRbMWFkpMtNpxq7UGYXsDCXG2ufCcKRcUFSTjRbENJRTODDo0bz/bznc9BZ8+ePfjKV76CU6dOeU+PlgiCALfbLVtx5+ru7kZVVZX3dk1NDQ4dOgSLxYKsrCy/PW4gSKM5cybGI9rk838WIgpSRQVJeGXnSZRWtFxwMwfRaIii6O2PpsW2If7i8zvqqlWrMHfuXPzzn/9EWlpaQP/S7t+/H1deeaX3trTQ+K677sKmTZsCVoc/DPa80e6CMKJQtCA7AUaDDvUdvahu7kZecozSJZFKVTfbUN/RC6NBh8Kc4NrlHMx8DjqVlZX4y1/+gry8PH/Uc1HLli07bxRJC+wuN3af6O95w5ROpC0RRj0Ksy0oq2xBSUULgw6NmTSaM3+SBZFGjvyPls+dkQsLC4dNH9H4HTjZjh6HG0kxJkxN4z+CRFrD87hIDiWcthoTnyPhgw8+iIcffhhnz57FjBkzztt9NXPmTNmKCxXexWX52jxQjSjUFRck4cl/HsPeE63oc7oRHqZXuiRSmT6nG3sHut3zEGPf+Bx0br31VgDA1772Ne/XBEHwLrLz52JkrRpcRR98fUCIaPzykqORZg5HQ2cf9ta08RM5+WxfTRv6nB6kxoYjPzn60j9AXj4HnZqaGn/UEbIarX04frYLggAs1ejJsUShThAEFBck4c2P6lBS3sygQz4bOm3FkX/f+Bx0Jk6c6I86Qpa0uGzmBDMsUUaFqyEifykaCDqlPGGdxqCU/XPGzOeg8+qrr170+3feeeeYiwlFXFxGFBoW5yVCrxNQ1dSN0+09yIhnV1sanTMdvahs6oZOAJbkcYmDr3wOOt/+9reH3XY6nejp6YHRaERkZCSDjg/cHhFllS0AuLiMSOvMEWG4PDMOB061o7SiBV8pVHeTUwocaTTn8sw4mCN5/JKvfN5e3t7ePuxPd3c3ysvLsWTJErzxxhv+qFGzPjndgc5eJ2LCDZiVEad0OUTkZ9LIbSm3mZMPBkf+2VB2LHwOOiPJz8/HU089dd5oD12c9OJdmp8Ig16W/xREFMSkoLOzqgVOt+cS9yYCXG4PdlRx5H88ZHt3NRgMOHPmjFyXCwmlQ/rnEJH2TZ9gRnxkGLrsLhyq61C6HFKBQ3Ud6OpzIS4yDDMmmJUuR5V8XqPz97//fdhtURTR0NCA559/HosXL5atMK3r7HF6/6HjKnqi0KDXCVian4S/Hz6DkvJmzJvE84ro4gZH/pOg13Fb+Vj4HHRuuummYbcFQUBSUhKuuuoq/PKXv5SrLs3bUdUCjwjkJ0cjPS5C6XKIKECKCvqDTmllMx5ZMVnpcijIDY78c7fVWPkcdDwezivLoaSiCQC3lROFGukN65PTnWjptiMx2qRwRRSs2mwOfFLfCYDvFeMxrjU6oihq8jRxfxNFcXAVPReXEYWU5NhwTE2LBQDsGGgvQTSSsspmiCIwJTUGybHhSpejWmMKOq+++ipmzJiBiIgIREREYObMmfjjH/8od22aVdHYjUarHeFhOs7RE4UgbjOn0eAHYnn4HHSeeeYZ3Hfffbj++uuxZcsWbNmyBddeey1WrVqF//3f//VHjZojTVstyEngKcZEIcgbdCqb4fFwVJzO5/GIKK0Y2FbOaatx8XmNznPPPYcXX3xxWAfkz3/+87jsssvw+OOP4zvf+Y6sBWqR9OLltnKi0DRnYjyijHq0dDtwtMGK6dw2TOc4dtaKlm47Io16zJ3Ikf/x8HlEp6GhAYsWLTrv64sWLUJDQ4MsRWlZj8OFfTVtADgcSRSqjAYdFub2L0ou4fQVjUB6XSzKTYDRwIay4+Hzs5eXl4ctW7ac9/XNmzcjPz9flqK0bM+JVjjcHkyIi0BOYpTS5RCRQooL+oMO1+nQSHhauXx8nrpav349br/9dpSWlnobBO7cuRMffPDBiAGIhvPOuU5OgiCw+RNRqOo/t+gzHDjVjq4+J2LCeVgj9eu2u7D/ZDsArs+Rg88jOrfeeiv27t2LxMREbN26FVu3bkViYiL27duHm2++2R81asrg4Wx88RKFsqyESExKiITLI2JXdavS5VAQ2V3dCpdHxMSESExM4Mj/ePk8ogMAc+bMwWuvvSZ3LZpX29qDmhYbDDoBi3ITlC6HiBRWXJCEk7tPobSiGSsuS1W6HAoSbCgrr1GP6Jw5cwaPPPIIrFbred/r7OzEd7/7XTQ2NspanNaUVPaP5lwxMZ7D1ETk3ZBQUtHM5qsE4JyGsgw6shh10HnmmWdgtVoRGxt73vfMZjO6urrwzDPPyFqc1pTyxUtEQyzISYBRr8Pp9l7UtNiULoeCwMnWHtS19SJML2BBDkf+5TDqoLNt27ZhvXPOdeedd+Kdd96RpSgtcrg82FXF5k9ENCjSaMC87HgA3GZO/aQPxPMmWRBlGtPqEjrHqINOTU0NsrKyLvj9jIwMnDx5Uo6aNOnAqXbYHG4kRBkxLe38UTEiCk1S41AGHQIGXwfcVi6fUQediIiIiwaZkydPIiIiQo6aNKm0cvDFq9NxWzkR9ZPW6ew50Yo+p1vhakhJdpcbuwd24HHkXz6jDjqFhYUXPbjz1Vdfxfz582UpSotKyrk+h4jONzklBimxJvQ5PfjoZJvS5ZCC9p9sR6/TjeQYE6akxihdjmaMOug88sgjeOWVV/DII48M213V2NiIhx9+GJs2bcIjjzzilyLVrqmrD0cb+nerLclPVLgaIgomgiB4p6/YJTm0DZ22YkNZ+Yw66Fx55ZV44YUX8PzzzyM9PR3x8fGwWCxIT0/HCy+8gOeeew5XXXWVP2tVrbKBbsgzJpiRGG1SuBoiCjZDt5lT6OLOXP/waUn3N7/5TXzuc5/Dli1bUFVVBVEUUVBQgC984QvIyMjwV42qN7g+h6M5RHS+JXmJ0AlARWM3Gjp7kWbmesdQc7azD8fPdkEQ+l8PJB+f965NmDAB3/nOd/xRiya5PeKQlJ6scDVEFIziIo2YlRmHj2s7UFrRjNvnXXiHK2mT9IF4VkYc4qOMClejLTz73c+O1HeivceJaJMBs7PilC6HiIIUt5mHNm4r9x8GHT+TRnMW5yUgTM+nm4hGJq3T2VHZApfbo3A1FEhuj4gdlWwo6y985/WzEk5bEdEozMqIgzkiDNY+Fw6f7lC6HAqgw6c70NnrRGy4AbMyzEqXozkMOn7U2evEx3UdALgQmYguTq8TvO0nSgZ2alJokPqsLc1PgoEj/7Lz+Rl97LHH8OGHH6Kvr88f9WjKrqoWuD0icpOikBEfqXQ5RBTkpGkLrtMJLdJCZE5b+YfPQWf37t244YYbEBcXh6VLl+JHP/oR3n//ffT29vqjPlXj4jIi8oW0IPmT0x1otzkUroYCoaPHgcMDI/9LOfLvFz4Hnffeew8dHR344IMPcP3112P//v245ZZbEBcXhyVLlvijRlUSRZHNn4jIJ6nmcExJjYEoAmVVnL4KBTuqWuAR+48CYf8k/xjTGfAGgwGLFy9GUlISLBYLYmJisHXrVhw/flzu+lSrqqkbZzr7YDToUJidoHQ5RKQSRQVJOH62CyXlzfj8rHSlyyE/k9bncB2n//g8ovO73/0OX/nKVzBhwgQsWrQI27Ztw5IlS7B//340N3NeWSJNWxVmWxBh1CtcDRGphTQCXFrZDFEUFa6G/EkUxSHrc7gz1198HtFZtWoVkpKS8PDDD+P+++9HdHS0P+pSvRJOWxHRGMydFI+IMD2au+w41tCFaemxSpdEflLe2IVGqx3hYTrMnRSvdDma5fOIzl//+lfccccdePPNN5GUlIRFixbhBz/4Ad5991309PT4o0bV6XO6sa+mDQCDDhH5xmTQY2Fu/3S39GmftEmatlqYk4DwMI78+4vPQeemm27CM888g4MHD+Ls2bP4wQ9+gPr6enzuc5+DxWLxR42qs+dEK+wuD9LN4chL5ogXEfnGu828nEFHy7itPDDGtBi5tbUVJSUl2L59O7Zv347PPvsM8fHxWLp0qdz1qdLQbeWCIChcDRGpjdSSYv+pNtjsLkSZxvRPNQWxHocLH9W0A2ALEn/z+W/PjBkzcOzYMcTHx6OoqAj33nsviouLMXPmTH/Up0rcVk5E4zEpIRJZlkjUtvVgd3Urlk9LUbokktmeE61wuD3ItEQgOzFK6XI0bUyLkYuLizF9+nR/1KN6p9t7UN1sg14nYFEetwsSke8EQUBRQSJe21OLkopmBh0N8m4rz+fIv7/5vEZn9erVmD59OhwOB8rLy+FyufxRl2qVDpxRMzuz/4A+IqKxkLYbc0GyNpXytPKA8Tno9Pb2YuXKlYiMjMRll12G2tpaAMCDDz6Ip556SvYC1aakogkAX7xEND4LcxNg0Ak41dqDky02pcshGZ1qtaGmxQaDTvDusCP/8TnoPProozh8+DC2b9+O8PBw79eXL1+OzZs3y1qc2jjdHuysagXAxWVEND7RJoO3twpHdbRFWsc5Z2I8YsI58u9vPgedrVu34vnnn8eSJUuGzStedtllqK6ulrU4tfm4tgPddhcsUUbMmGBWuhwiUjlp+orbzLWlZGCJQ/FkfiAOBJ+DTnNzM5KTz29VbbPZQn5BlTRttSQvETpdaD8XRDR+0vlHu6pbYXe5Fa6G5OBwebC7uj/oSKfVk3/5HHTmzp2Lf/7zn97bUrj5/e9/j4ULF8pXmQpJC5G5PoeI5DAtLRZJMSb0Ot04cLJd6XJIBgdOtcPmcCMx2oRpaTzeIxB83l7+05/+FNdddx2OHj0Kl8uFX//61zh69Ch27dqFkpISf9SoCi3ddnxa3wkAWMpTaIlIBoIgYGl+Iv56sB4lFc1sWaEB3oay+Rz5DxSfR3SWLFmCQ4cOweVyYcaMGXj33XeRnJyM3bt3Y86cOf6oURV2DGwVnJYWi+SY8Evcm4hodLzHQVRwnY4WeBvKcn1OwPg8onPkyBFMnz4dL7300nnf27p1K2666SY56lKdEr54icgPluYnQRCA42e70GjtQ0osP0ipVVNXH442WCEI/Ws5KTB8HtFZsWIFampqzvv6W2+9hTvuuEOWoi7mhRdewKRJkxAeHo7CwkLs27fP7495KR6P6E3pXFxGRHKyRBkxc2AXZylHdVStbGAd54wJZiREmxSuJnT4HHS+/vWvY/ny5Th79qz3a5s3b8add96JTZs2yVnbeTZv3ow1a9Zg3bp1OHjwIGbNmoUVK1agqanJr497KUcbrGi1ORBl1GPOxHhFayEi7eH0lTaU8BxERfgcdNavX4/rr78ey5cvR1tbG15//XXcc889ePXVV/HFL37RHzV6PfPMM7j33ntxzz33YNq0afjNb36DyMhIbNy40a+PeynSi3dhbiKMBp+fUiKii5IakJZVtsDtERWuhsbC7RFRNtD4kQ1lA2tM78rPPfccZs2ahQULFuDee+/FG2+8gVtvvVXu2oZxOBw4cOAAli9f7v2aTqfD8uXLsXv37hF/xm63w2q1DvvjD1yfQ0T+dHlmHGLCDejsdeLw6Q6ly6ExOFLfifYeJ2JMBlyeGad0OSFlVIuR//73v5/3tVtuuQVlZWX48pe/DEEQvPf5/Oc/L2+FA1paWuB2u5GSMvwU35SUFBw/fnzEn9mwYQPWr1/vl3qG+vL8TKTEhmMZUzoR+YFBr8OSvET8+8hZlFY044osTpGrjfSBeHFeIsL0HPkPpFEFnYvtpNq4caN36kgQBLjdwdO9c+3atVizZo33ttVqRWZmpuyPc/PsDNw8O0P26xIRSYoLkrxB56HlBUqXQz7itnLljCroeDwef9dxSYmJidDr9WhsbBz29cbGRqSmpo74MyaTCSYTV7YTkfpJ6zoO1XWgs8cJcyQPg1SLzh4nDtb2d7bm+pzAU834mdFoxJw5c/DBBx94v+bxePDBBx+E/NETRKR96XERyE+OhkcEdlS1KF0O+WBndQs8IpCXHI0JcRFKlxNyRh10du/ejXfeeWfY11599VVkZ2cjOTkZ3/jGN2C322UvcKg1a9bgpZdewh/+8AccO3YM9913H2w2G+655x6/Pi4RUTAY3GaubEsN8k0pt5UratRB54knnsBnn33mvf3pp59i5cqVWL58OR599FH84x//wIYNG/xSpOT222/H008/jcceewyXX345Dh06hG3btp23QJmISIuKhvTTEUVuM1cDURQHz7di0FHEqIPOoUOHcPXVV3tvv/nmmygsLMRLL72ENWvW4Nlnn8WWLVv8UuRQDzzwAE6dOgW73Y69e/eisLDQ749JRBQM5mdbEB6mQ6PVjorGbqXLoVGoaupGQ2cfTAYdCrMtSpcTkkYddNrb24eNnJSUlOC6667z3p43bx7q6urkrY6IiLzCw/QozE4AwOkrtZBGcwpzEhAeple4mtA06qCTkpLiPePK4XDg4MGDWLBggff7XV1dCAvjLgAiIn+S1nmUVnBBshrw2AfljTroXH/99Xj00UdRVlaGtWvXIjIyEkuXLvV+/5NPPkFubq5fiiQion7SOo99NW3ocbgUroYuptfhxt6aNgBAcQFPK1fKqIPO//zP/8BgMKC4uBgvvfQSXnrpJRiNRu/3N27ciGuuucYvRRIRUb/cpChMiIuAw+3B3hNtSpdDF7GnphUOlwcT4iKQmxStdDkha1QNA4H+hn2lpaXo7OxEdHQ09Prhc41//vOfER3N/5BERP4kCAKKJyfh9b21KKloxpVTkpUuiS6g1LvbKhGCIChcTejyuWGg2Ww+L+QAgMViGTbCQ0RE/lGUP7jNnIIX1+cEB9V0RiYion6L8hJg0AmoabGhtrVH6XJoBHVtPTjRbINeJ2BRHtfnKIlBh4hIZWLDw7wnmJdUclQnGJUO/He5IisOseHckawkBh0iIhWSTsEu5fRVUCop57RVsGDQISJSIekNdFdVCxwuj8LV0FBOtwe7qlsB8NiHYMCgQ0SkQtPSYpEQZYTN4cbB2naly6EhDp5qR7fdBUuUEdPTzUqXE/IYdIiIVEinE4Yd8knBQ1qfszQ/ETodt5UrjUGHiEiliga67UrrQSg4cFt5cGHQISJSqaUD/XSONljR1NWncDUEAC3ddhyptwIY/O9DymLQISJSqcRoE6ZPiAUAlPGQz6BQNjBtdVl6LJJiTApXQwCDDhGRqnlPM2c/naDAbeXBh0GHiEjFigv6z7oqq2yBxyMqXE1o83hElFX2j6xxW3nwYNAhIlKx2VlxiDYZ0GZz4MiZTqXLCWmfnbGi1eZAtMng7VxNymPQISJSsTC9DovzEgBw95XSpOnDhbkJMBr49hos+F+CiEjl2E8nOHB9TnBi0CEiUrmigW3MH9d1oLPXqXA1ocna58SBgQ7VDDrBhUGHiEjlMi2RyEmKgtsjYlcVt5krYVdVK9weETmJUci0RCpdDg3BoENEpAHcZq4sadqQu62CD4MOEZEGSEGnpLwZosht5oEkiiJKeexD0GLQISLSgMLs/p0+Zzr7UN3crXQ5IaW62Yb6jl4YDToU5liULofOwaBDRKQBEUY9CrP732S3c5t5QEnTVvMnWRBpNChcDZ2LQYeISCOKuc1cEZy2Cm4MOkREGiG90e6raUOf061wNaGhz+nGnhOtALgQOVgx6BARaURecjTSzOGwuzzeN1/yr301bbC7PEiNDUdBSrTS5dAIGHSIiDRCEITBbeYV7KcTCIPbyhMhCILC1dBIGHSIiDRkcJ1Ok8KVhIbB9TnJCldCF8KgQ0SkIYvyEqHXCahutuF0e4/S5WhafUcvKpu6oROAJXmJSpdDF8CgQ0SkIeaIMMzOjAPA6St/k0ZzLs+MgzkyTOFq6EIYdIiINKaI01cBwWkrdWDQISLSGGmdzq6qVjjdHoWr0SaX24MdAweoFhVw2iqYMegQEWnM9AlmxEeGocvuwse1HUqXo0mH6jrQ1edCXGQYZmbEKV0OXQSDDhGRxuh1ApbmS9vM2SXZH6Rt5UsGFn9T8GLQISLSIB4H4V8lPPZBNRh0iIg0aOnAupFP6zvR0m1XuBptae2249P6TgA89kENGHSIiDQoOSYc09JiAQA7KrnNXE47qlogisCU1BikxIYrXQ5dAoMOEZFGFXH6yi+801aTOZqjBgw6REQaJa0fKatshscjKlyNNng8orcRY3E+g44aMOgQEWnUnInxiDLq0dLtwNEGq9LlaMKxs1a0dNsRadRjzqR4pcuhUWDQISLSKKNBh4W5/YuSOX0lD+l5XJiTAJNBr3A1NBoMOkREGiatI2HQkUdJOdfnqA2DDhGRhknrSA6eakdXn1PhatSt2+7CgVPtAIAirs9RDQYdIiINy0qIRHZiFFweEbuqW5UuR9V2VbXA5RExMSESkxKjlC6HRolBh4hI44ryuU5HDqWV7IasRgw6REQaJ60nKa1ohihym/lYiKKI7QPrczhtpS4MOkREGrcgJwFGvQ6n23txosWmdDmqVNNiw+n2XoTpBSzMTVC6HPIBgw4RkcZFGg2Yl93f84WnmY+N9LzNnWhBlMmgcDXkCwYdIqIQwNPMx4fHPqgXgw4RUQiQzr3ac6IVfU63wtWoS5/TjT0n2gBwfY4aMegQEYWAySkxSIk1oc/pwUcn25QuR1X2n2xHr9ONpBgTpqbFKF0O+YhBh4goBAiC4B2NkLr70uhI28qL8pMgCILC1ZCvGHSIiEKEd5t5JYOOL3jsg7qpJuj85Cc/waJFixAZGYm4uDilyyEiUp0leYnQCUBFYzfOdPQqXY4qNHT2oryxC4IALM1LVLocGgPVBB2Hw4EvfvGLuO+++5QuhYhIleIijZiVGQcAKOOozqiUVbQAAGZmxCE+yqhwNTQWqmkGsH79egDApk2blC2EiEjFiguS8HFtB97+uB7miLBxX08nCCjMSZDlWnLp7HVi74lWeGToAv32x/UAeOyDmqkm6IyF3W6H3W733rZarQpWQ0SkvKKCJPzq/UrsOdHm3TI9XsunJuP3d82T5VpyeHjLYbx/rFHWaxYXcNpKrTQddDZs2OAdCSIiIuDyjDjcvWgSjtR3jvtablHEx7UdKK1oQY/DhUij8m8pPQ6Xt4vx7Kw46GXYJTUtPRazM+PHfR1ShqKvykcffRQ/+9nPLnqfY8eOYcqUKWO6/tq1a7FmzRrvbavViszMzDFdi4hIC3Q6AY9//jJZriWKIpb+/EOcbu/FnhOtuGpKiizXHY+9J9rgcHswIS4Cf71vEbeDk7JB5+GHH8bdd9990fvk5OSM+fomkwkmk2nMP09ERBcmCAKKCpLw+t5alJQ3B0XQkY5qKCpgzxvqp2jQSUpKQlISF3gREalV8UDQKa1sUboUAIOHb3LxMEmUn1AdpdraWrS1taG2thZutxuHDh0CAOTl5SE6OlrZ4oiIQtSi3AQYdAJqWmyobe1BVkKkYrXUtfXgRIsNBp2ARXkJitVBwUU1fXQee+wxzJ49G+vWrUN3dzdmz56N2bNnY//+/UqXRkQUsmLCw3DFxP6FuiUK9+aRpq2uyIpHbHjwbHcnZakm6GzatAmiKJ73Z9myZUqXRkQU0qRpIqXP0JKCDo9qoKFUE3SIiCg4SUFnd3ULHC6PIjU4XB7srm4FAO/hpUQAgw4REY3TtLRYJEYbYXO4ceBUuyI1HKxtR7fdhYQoIy5Lj1WkBgpODDpERDQuOp2ApQOjKNL0UaBJj7s0PxE6HbeV0yAGHSIiGjdp+qpUoaBTyvU5dAEMOkRENG5L8xMhCMDRBiuauvoC+tjNXXZ8dsY6UAeDDg3HoENEROOWEG3C9HQzAKCsIrDNA8sGtrVPnxCLxGh2w6fhGHSIiEgW3m3mAZ6+KmE3ZLoIBh0iIpJF0UDQKKtshtsjBuQxPR4RZQPHT3BbOY2EQYeIiGQxOysOMSYD2nucOFLfGZDHPHKmE202B6JNBm+HZqKhGHSIiEgWYXqd94ypQE1fSd2YF+UmIEzPtzQ6H18VREQkm+KCZACB22ZeWslt5XRxDDpERCSbooJEAMDHdR3o7HX69bGsfU4crO3of1yuz6ELYNAhIiLZZMRHIjcpCm6PiF1V/t1mvquqBW6PiJykKGRaIv36WKReDDpERCQrafrK3+t0uK2cRoNBh4iIZCVNX5VUNEMU/bPNXBRF70LkIgYduggGHSIiktWCnASYDDo0dPahqqnbL49R3dyNM519MBp0WJCd4JfHIG1g0CEiIlmFh+kxP9sCwH/TV9sHRnMKsy2IMOr98hikDQw6REQkO38fB1E60A2Z63PoUhh0iIhIdssG+trsrWlDr8Mt67X7nG7sPdEKgEGHLo1Bh4iIZJebFI10czgcLg/21rTKeu29NW2wuzxIM4cjLzla1muT9jDoEBGR7ARB8HYrlnv6StptVVyQBEEQZL02aQ+DDhER+YXUrVj2oFPR1H99TlvRKDDoEBGRXyzKS4ReJ+BEsw11bT2yXPN0ew+qm23Q6wQszkuU5ZqkbQw6RETkF+aIMMzOjAMwePjmeJVW9O+2ujwzDuaIMFmuSdrGoENERH4j7YqS6zTzUh77QD5i0CEiIr+RFiTvrGqF0+0Z17Wcbg92VrF/DvmGQYeIiPxmeroZligjuu0ufFzbMa5rHarrQJfdhfjIMEyfYJanQNI8Bh0iIvIbnU7A0nzpkM+mcV1L2la+ND8Jeh23ldPoMOgQEZFfybXNXPp5bisnXzDoEBGRXy0t6B/ROVJvRUu3fUzXaOm249P6TgBAUT63ldPoMegQEZFfJceEY1paLACgbIzbzHcMHOI5NS0WybHhstVG2segQ0REfiftvpL64PiK28pprBh0iIjI74b20/F4RJ9+1uMRvQ0HGXTIVww6RETkd1dkxSPKqEerzYGjDVaffvZogxUt3Q5EGfWYMzHeTxWSVjHoEBGR3xkNOizKk7aZ+7ZOR7r/wtxEGA182yLf8BVDREQBIW0Ll/rhjFaJd30Od1uR7xh0iIgoIIoH+ukcrG2Htc85qp/p6nPi4Kn2/p8vSPZbbaRdDDpERBQQWQmRyE6MgssjYldV66h+Zld1K1weEZMSIpGVEOnnCkmLGHSIiChgvLuvRtlPh9vKabwYdIiIKGCKh6zTEcWLbzMXRXFwfc5kBh0aGwYdIiIKmMIcC4x6Heo7enGixXbR+9a02HC6vRdGvQ4LchICVCFpDYMOEREFTKTRgPnZFgCX3n0ljebMy45HpNHg99pImxh0iIgooIoKRtdPx3taeT6nrWjsGHSIiCigpG3ie2ta0ed0j3ifPqcbe07078zi+hwaDwYdIiIKqIKUaKTGhqPP6cG+mrYR7/PRyTb0OT1IiTVhckpMgCskLWHQISKigBIEwTt9VXqB6avSIdNWgiAErDbSHgYdIiIKOGn66kLrdLitnOTCoENERAG3JC8ROgGobOrGmY7eYd9r6OxFRWM3dEL//YjGg0GHiIgCzhwZhssz4wCcP30l3Z6VGYe4SGOgSyONYdAhIiJFeE8zPyfocFs5yYlBh4iIFCEdB7GjqgUutwcA4HJ7sKOypf/7XJ9DMmDQISIiRczMiENcZBi6+lw4VNcBADh8ugPWPhfMEWGYlRGnaH2kDQw6RESkCL1O8C42ltbllFT0j+YsyU+EXsdt5TR+DDpERKSY4nPW6Xi3lRdw2orkwaBDRESKkRYkf1LfiermbnxyuqP/61yITDJh0CEiIsWkxIZjSmoMRBHY8K/jEEVgSmoMUs3hSpdGGqGKoHPy5EmsXLkS2dnZiIiIQG5uLtatWweHw6F0aURENE7SNNX7xxoBDI7yEMnBoHQBo3H8+HF4PB789re/RV5eHo4cOYJ7770XNpsNTz/9tNLlERHROBQXJOG3pSeG3SaSiyqCzrXXXotrr73WezsnJwfl5eV48cUXGXSIiFRuzqR4RITp0et0IyJMj7mT4pUuiTREFVNXI+ns7ITFYrnofex2O6xW67A/REQUXEwGPRblJgAAFuYmwGTQK1wRaYkqg05VVRWee+45fPOb37zo/TZs2ACz2ez9k5mZGaAKiYjIFyuXZCPLEomVS7KVLoU0RhBFUVTqwR999FH87Gc/u+h9jh07hilTpnhv19fXo7i4GMuWLcPvf//7i/6s3W6H3W733rZarcjMzERnZydiY2PHVzwREREFhNVqhdlsHtP7t6JBp7m5Ga2trRe9T05ODozG/tNrz5w5g2XLlmHBggXYtGkTdDrfBqTG80QRERGRMsbz/q3oYuSkpCQkJY1udX19fT2uvPJKzJkzB6+88orPIYeIiIhCjyp2XdXX12PZsmWYOHEinn76aTQ3N3u/l5qaqmBlREREFMxUEXTee+89VFVVoaqqChkZGcO+p+DMGxEREQU5Vcz/3H333RBFccQ/RERERBeiiqBDRERENBYMOkRERKRZDDpERESkWQw6REREpFkMOkRERKRZDDpERESkWQw6REREpFkMOkRERKRZDDpERESkWao4AkIuUidlq9WqcCVEREQ0WtL79lhORAipoNPV1QUAyMzMVLgSIiIi8lVXVxfMZrNPPyOIIXRglMfjwZkzZxATEwNBEGS7rtVqRWZmJurq6hAbGyvbdeni+Lwrg8+7Mvi8K4PPuzLOfd5FUURXVxfS09Oh0/m26iakRnR0Ot15p5/LKTY2ln8RFMDnXRl83pXB510ZfN6VMfR593UkR8LFyERERKRZDDpERESkWQw6MjCZTFi3bh1MJpPSpYQUPu/K4POuDD7vyuDzrgw5n/eQWoxMREREoYUjOkRERKRZDDpERESkWQw6REREpFkMOkRERKRZDDoyeOGFFzBp0iSEh4ejsLAQ+/btU7okTXv88cchCMKwP1OmTFG6LM0pLS3FDTfcgPT0dAiCgK1btw77viiKeOyxx5CWloaIiAgsX74clZWVyhSrIZd63u++++7zXv/XXnutMsVqxIYNGzBv3jzExMQgOTkZN910E8rLy4fdp6+vD6tXr0ZCQgKio6Nx6623orGxUaGKtWE0z/uyZcvOe72vWrXKp8dh0BmnzZs3Y82aNVi3bh0OHjyIWbNmYcWKFWhqalK6NE277LLL0NDQ4P2zY8cOpUvSHJvNhlmzZuGFF14Y8fs///nP8eyzz+I3v/kN9u7di6ioKKxYsQJ9fX0BrlRbLvW8A8C111477PX/xhtvBLBC7SkpKcHq1auxZ88evPfee3A6nbjmmmtgs9m89/nOd76Df/zjH/jzn/+MkpISnDlzBrfccouCVavfaJ53ALj33nuHvd5//vOf+/ZAIo3L/PnzxdWrV3tvu91uMT09XdywYYOCVWnbunXrxFmzZildRkgBIL799tve2x6PR0xNTRV/8YtfeL/W0dEhmkwm8Y033lCgQm0693kXRVG86667xBtvvFGRekJFU1OTCEAsKSkRRbH/tR0WFib++c9/9t7n2LFjIgBx9+7dSpWpOec+76IoisXFxeK3v/3tcV2XIzrj4HA4cODAASxfvtz7NZ1Oh+XLl2P37t0KVqZ9lZWVSE9PR05ODu644w7U1tYqXVJIqampwdmzZ4e99s1mMwoLC/naD4Dt27cjOTkZkydPxn333YfW1lalS9KUzs5OAIDFYgEAHDhwAE6nc9jrfcqUKcjKyuLrXUbnPu+SP/3pT0hMTMT06dOxdu1a9PT0+HTdkDrUU24tLS1wu91ISUkZ9vWUlBQcP35coaq0r7CwEJs2bcLkyZPR0NCA9evXY+nSpThy5AhiYmKULi8knD17FgBGfO1L3yP/uPbaa3HLLbcgOzsb1dXV+MEPfoDrrrsOu3fvhl6vV7o81fN4PHjooYewePFiTJ8+HUD/691oNCIuLm7Yffl6l89IzzsAfOUrX8HEiRORnp6OTz75BN///vdRXl6Ov/71r6O+NoMOqc51113n/f8zZ85EYWEhJk6ciC1btmDlypUKVkbkf1/60pe8/3/GjBmYOXMmcnNzsX37dlx99dUKVqYNq1evxpEjR7juL8Au9Lx/4xvf8P7/GTNmIC0tDVdffTWqq6uRm5s7qmtz6mocEhMTodfrz1t539jYiNTUVIWqCj1xcXEoKChAVVWV0qWEDOn1zde+8nJycpCYmMjXvwweeOABvPPOO/jwww+RkZHh/XpqaiocDgc6OjqG3Z+vd3lc6HkfSWFhIQD49Hpn0BkHo9GIOXPm4IMPPvB+zePx4IMPPsDChQsVrCy0dHd3o7q6GmlpaUqXEjKys7ORmpo67LVvtVqxd+9evvYD7PTp02htbeXrfxxEUcQDDzyAt99+G//5z3+QnZ097Ptz5sxBWFjYsNd7eXk5amtr+Xofh0s97yM5dOgQAPj0eufU1TitWbMGd911F+bOnYv58+fjV7/6FWw2G+655x6lS9OsRx55BDfccAMmTpyIM2fOYN26ddDr9fjyl7+sdGma0t3dPexTU01NDQ4dOgSLxYKsrCw89NBDePLJJ5Gfn4/s7Gz8+Mc/Rnp6Om666SblitaAiz3vFosF69evx6233orU1FRUV1fje9/7HvLy8rBixQoFq1a31atX4/XXX8ff/vY3xMTEeNfdmM1mREREwGw2Y+XKlVizZg0sFgtiY2Px4IMPYuHChViwYIHC1avXpZ736upqvP7667j++uuRkJCATz75BN/5zndQVFSEmTNnjv6BxrVni0RRFMXnnntOzMrKEo1Gozh//nxxz549SpekabfffruYlpYmGo1GccKECeLtt98uVlVVKV2W5nz44YcigPP+3HXXXaIo9m8x//GPfyympKSIJpNJvPrqq8Xy8nJli9aAiz3vPT094jXXXCMmJSWJYWFh4sSJE8V7771XPHv2rNJlq9pIzzcA8ZVXXvHep7e3V7z//vvF+Ph4MTIyUrz55pvFhoYG5YrWgEs977W1tWJRUZFosVhEk8kk5uXlid/97nfFzs5Onx5HGHgwIiIiIs3hGh0iIiLSLAYdIiIi0iwGHSIiItIsBh0iIiLSLAYdIiIi0iwGHSIiItIsBh0iIiLSLAYdIiIi0iwGHSLyK0EQsHXrVkVrEEUR3/jGN2CxWCAIgve8HCLSPgYdIhrR3XffrZlzq7Zt24ZNmzbhnXfeQUNDA6ZPn37efbZv3w5BECAIAnQ6HcxmM2bPno3vfe97aGhoGHbfxx9/HIIgYNWqVcO+fujQIQiCgJMnT553/RUrVkCv1+Ojjz6S9Xcjootj0CEizZNOt1+0aBFSU1NhMFz4POPy8nKcOXMGH330Eb7//e/j/fffx/Tp0/Hpp58Ou194eDhefvllVFZWXvLxa2trsWvXLjzwwAPYuHHjuH8fIho9Bh0iGpVly5bhW9/6Fr73ve/BYrEgNTUVjz/++LD7VFZWoqioCOHh4Zg2bRree++9865TV1eH2267DXFxcbBYLLjxxhu9IyDHjx9HZGQkXn/9de/9t2zZgoiICBw9evSCtZWUlGD+/PkwmUxIS0vDo48+CpfLBaB/ZOrBBx9EbW0tBEHApEmTLvp7JicnIzU1FQUFBfjSl76EnTt3IikpCffdd9+w+02ePBlXXnklfvjDH170egDwyiuv4HOf+xzuu+8+vPHGG+jt7b3kzxCRPBh0iGjU/vCHPyAqKgp79+7Fz3/+czzxxBPeMOPxeHDLLbfAaDRi7969+M1vfoPvf//7w37e6XRixYoViImJQVlZGXbu3Ino6Ghce+21cDgcmDJlCp5++mncf//9qK2txenTp7Fq1Sr87Gc/w7Rp00asqb6+Htdffz3mzZuHw4cP48UXX8TLL7+MJ598EgDw61//Gk888QQyMjLQ0NDg89RRREQEVq1ahZ07d6KpqWnY95566im89dZb2L9//wV/XhRFvPLKK/jqV7+KKVOmIC8vD3/5y198qoGIxo5Bh4hGbebMmVi3bh3y8/Nx5513Yu7cufjggw8AAO+//z6OHz+OV199FbNmzUJRURF++tOfDvv5zZs3w+Px4Pe//z1mzJiBqVOn4pVXXkFtbS22b98OALj//vuxZMkSfPWrX8Xdd9+NefPm4cEHH7xgTf/3f/+HzMxMPP/885gyZQpuuukmrF+/Hr/85S/h8XhgNpsRExMDvV6P1NRUJCUl+fx7T5kyBQDOW3tzxRVX4Lbbbjsv0A31/vvvo6enBytWrAAAfPWrX8XLL7/scw1ENDYXnqgmIjrHzJkzh91OS0vzjnIcO3YMmZmZSE9P935/4cKFw+5/+PBhVFVVISYmZtjX+/r6UF1d7b29ceNGFBQUQKfT4bPPPoMgCBes6dixY1i4cOGw+yxevBjd3d04ffo0srKyfP9FzyGKIgCMWMeTTz6JqVOn4t1330VycvJ539+4cSNuv/1277qgL3/5y/jud7+L6upq5Obmjrs2Iro4Bh0iGrWwsLBhtwVBgMfjGfXPd3d3Y86cOfjTn/503veGjrQcPnwYNpsNOp0ODQ0NSEtLG3vRMjh27BgAjLi+Jzc3F/feey8effTR80Zq2tra8Pbbb8PpdOLFF1/0ft3tdmPjxo34yU9+4te6iYhBh4hkMnXqVNTV1Q0LJnv27Bl2nyuuuAKbN29GcnIyYmNjR7xOW1sb7r77bvzwhz9EQ0MD7rjjDhw8eBAREREXfNy33noLoih6R1x27tyJmJgYZGRkjPv36u3txe9+9zsUFRVdcNrrscceQ25uLt58881hX//Tn/6EjIyM8/oIvfvuu/jlL3+JJ554Anq9ftw1EtGFcY0OEcli+fLlKCgowF133YXDhw+jrKzsvB1Jd9xxBxITE3HjjTeirKwMNTU12L59O771rW/h9OnTAIBVq1YhMzMTP/rRj/DMM8/A7XbjkUceueDj3n///airq8ODDz6I48eP429/+xvWrVuHNWvWQKfz/Z+4pqYmnD17FpWVlXjzzTexePFitLS0DBuROVdKSgrWrFmDZ599dtjXX375ZXzhC1/A9OnTh/1ZuXIlWlpasG3bNp/rIyLfMOgQkSx0Oh3efvtt9Pb2Yv78+fj6179+3tRMZGQkSktLkZWVhVtuuQVTp07FypUr0dfXh9jYWLz66qv417/+hT/+8Y8wGAyIiorCa6+9hpdeegn//ve/R3zcCRMm4F//+hf27duHWbNmYdWqVVi5ciV+9KMfjen3mDx5MtLT0zFnzhw89dRTWL58OY4cOXLBXV+SRx55BNHR0d7bBw4cwOHDh3Hrrbeed1+z2Yyrr76ai5KJAkAQpVV2RERERBrDER0iIiLSLAYdIiIi0iwGHSIiItIsBh0iIiLSLAYdIiIi0iwGHSIiItIsBh0iIiLSLAYdIiIi0iwGHSIiItIsBh0iIiLSLAYdIiIi0qz/H3lasfY9Y3nYAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "SEQ_O = Seq('AGTGACGTAGCACATCCACGAGGTG')\n",
    "\n",
    "OUTPUT_PYTHON, skew_counts = minimum_skew(SEQ_O)\n",
    "print('Python: The minimum skew is at index:', OUTPUT_PYTHON)\n",
    "\n",
    "# Plot the skew_counts to check them out\n",
    "import matplotlib.pyplot as plt\n",
    "plt.plot(skew_counts)\n",
    "plt.ylabel('Skew Counts')\n",
    "_=plt.xlabel('Index of DNA')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FHE: The minimum skew is at index:\n",
      "|  Compiling  : 4.39 s  |\n",
      "|  Encrypting  : 0.71 s  |\n",
      "|  Running  : 27.24 s  |\n",
      "|  Decrypting  : 0.00 s  |\n",
      "==> Result : 18 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "def process_sequence(seq):\n",
    "    return minimum_skew(seq)[0] # return only the minimum value for fhe\n",
    "\n",
    "compute_fhe_output([SEQ_O], process_sequence, seq_output = False,\n",
    "                   seq_interface= SeqInterface(Alphabets.DNA_RNA),\n",
    "                   description='FHE: The minimum skew is at index:')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"section-iv\"></a>\n",
    "## IV. Sorting sequences\n",
    "\n",
    "<a id=\"section-iv-1\"></a>\n",
    "### Comparing two sequences:\n",
    "\n",
    "As we already saw, <span style=\"color:green\">**Seq**</span> and <span style=\"color:orange\">**FheSeq**</span> classes implement some operators such as **len(**seq**)**, seq**[i:j]**,  or s1**==**s2. They also implement the comparison operators s1 **>** s2, s1 **>=** s2 and their opposite s1 **<** s2, s1 **<=** s2.  \n",
    "\n",
    "For the <span style=\"color:green\">**Seq**</span> class, a sequence **A** is lower than another sequence **B** (noted *A<B*) if **A** is alphabetically lower than **B** (and not equal to B). The first character is compared first:  \n",
    "- If *A[0] < B[0]* then *A < B*\n",
    "- Else, if *A[0] > B[0]* then *A > B*\n",
    "- Else, *A[0] = B[0]*, so the second character is compared, and so on.  \n",
    "\n",
    "For instance, `Seq('AB')` is lower than `Seq('AC')`. Also, `Seq('AA')` is lower than the `Seq('AAA')` because the empty character at the end is considered alphabetically lower than all other characters. The <span style=\"color:orange\">**FheSeq**</span> class works with the same principle. Comparing two sequences in alphabetical order is sometimes used inside bigger bioinformatics algorithms. If you need element wise comparison between <span style=\"color:orange\">**FheSeq**</span> objects instead, use `A.to_array() > B.to_array()`.\n",
    "\n",
    "<a id=\"section-iv-2\"></a>\n",
    "### Ranking a list of sequences:\n",
    "\n",
    "( difficulty level: *medium* )\n",
    "\n",
    "With such comparison operators, we are able to sort a list of sequences. On clear data, it is possible to use `sorted(sequence_list)` to sort a list of sequences, but in FHE, this is impossible for sequences of different lenghts. Indeed, sequences of different length are represented as python objects that are not directly integer arrays, and we cannot set a list of python objects with an encrypted index. For now, let's just compute a rank for each sequence object in the list. \n",
    "\n",
    "To do so, we can build a comparison matrix. Suppose our list of sequences is the following : `seqlist = ['CAA', 'AAB', 'CAA', 'AA', 'AAA']`, and we want an ascending ranking. In our comparison matrix, we will compute `comp_matrix[i,j] = seqlist[i]<seqlist[j]` when `i<j`(upper half), and `seqlist[i]>=seqlist[j]` equal to `1-comp_matrix[i,j]` when `j<i` (lower half).  \n",
    "\n",
    "Then, we can sum up the results on each row to get how many other sequences are lower than the sequence of the row, which gives exactly the ranking of this sequence in the list. Note that even though two identical sequences are provided, they get each a different rank, which is preferable (this is because we used `seqlist[i]>=seqlist[j]` instead of `seqlist[i]>seqlist[j]`).\n",
    "\n",
    "<div>\n",
    "<img src=\"https://rcd-media.com/docs/fhe/tableau_sort.png\" width=\"750\"/>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rank_sequences(seqlist, ascending=True):\n",
    "    \"\"\"\n",
    "    Returns an array with the rank of each sequence in the list\n",
    "    \"\"\"\n",
    "    n=len(seqlist)\n",
    "    \n",
    "    # Create a comparison matrix\n",
    "    comp_matrix = fhe.zeros((n,n)) # this works in regular python as well\n",
    "    \n",
    "    # Fill the matrix with comparison results\n",
    "    for i in range(0, n):\n",
    "        for j in range(i+1, n):\n",
    "            # Compute upper half of the matrix with seqlist[i]<seqlist[j]\n",
    "            if ascending:\n",
    "                comp_matrix[i,j] = (seqlist[i]<seqlist[j])\n",
    "            else:\n",
    "                comp_matrix[i,j] = (seqlist[i]>=seqlist[j])\n",
    "                \n",
    "            # Then fill the contrary to the lower half \n",
    "            comp_matrix[j,i] = 1-comp_matrix[i,j]\n",
    "            \n",
    "            # Note: 1-comp_matrix[i,j] is the same as seqlist[j]>=seqlist[i] (for ascending)\n",
    "            # Using seqlist[j]>=seqlist[i] instead of seqlist[j]>seqlist[i]\n",
    "            # will return a clear order, with different ranks even for\n",
    "            # equal sequences. Also, this is faster.\n",
    "            \n",
    "    # now sum up each row to get a sorting rank for the sequences\n",
    "    return np.sum(comp_matrix, axis=0)\n",
    "            \n",
    "def process_sequences(*seqlist):\n",
    "    return rank_sequences(seqlist, ascending = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's test this algorithm on the example sequence list:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sorted list: [Seq('AA'), Seq('AAA'), Seq('AAB'), Seq('CAA'), Seq('CAA')]\n",
      "\n",
      "The ranks of sequences are : [4 2 3 0 1]\n"
     ]
    }
   ],
   "source": [
    "# Test the sequence list of the example\n",
    "SEQ1 = Seq('AA')\n",
    "SEQ2 = Seq('AAA') # as mentioned earlier, SEQ2 is alphabetically greater than SEQ1\n",
    "SEQ3 = Seq('AAB')\n",
    "SEQ4 = Seq('CAA')\n",
    "SEQ5 = Seq('CAA') # same a SEQ4, will be ranked after it\n",
    "\n",
    "# Create a shuffled list, like in the example\n",
    "seqlist = [SEQ5, SEQ3, SEQ4, SEQ1, SEQ2]\n",
    "# Note: the expected ranks are thus: [5 3 4 1 2], or equally: [4 2 3 0 1]\n",
    "\n",
    "# Test sorted in python to compare (remember this cannot work in FHE):\n",
    "print('Sorted list:', sorted(seqlist))\n",
    "\n",
    "# Test our algorithm in python\n",
    "OUTPUT_PYTHON = process_sequences(*seqlist)\n",
    "print(\"\\nThe ranks of sequences are :\", OUTPUT_PYTHON)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FHE: The ranks of sequences are:\n",
      "|  Compiling  : 1.63 s  |\n",
      "|  Encrypting  : 0.07 s  |\n",
      "|  Running  : 0.35 s  |\n",
      "|  Decrypting  : 0.00 s  |\n",
      "==> Result : [4 2 3 1 0] \n",
      "\n"
     ]
    }
   ],
   "source": [
    "compute_fhe_output(seqlist, process_sequences, seq_output = False,\n",
    "                   seq_interface = SeqInterface(\"ABCD\"),\n",
    "                   description='FHE: The ranks of sequences are:')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"section-iv-3\"></a>\n",
    "### Sorting sequences of same length:\n",
    "\n",
    "( difficulty level: *hard* )\n",
    "\n",
    "We saw that in FHE, it is impossible to sort a list for sequences of different lengths. But for sequences of identical lengths (or with adjusted length as we will see), there is a very computationally intensive way of proceeding.  \n",
    "\n",
    "When given a list `[a,b]` of two encrypted numbers **a** and **b**, we can sort them using the formula `sortedlist  = [ a*(a<b) + b*(b<a), a*(a>b) + b*(b>a) ]`. This formula uses what we can call a *boolean sum*. For two numbers, this is already computationally intensive, but it works. For a bigger list of numbers `numlist`, the same applies. Let's see how this can be done, using python for now:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sorted : [0, 2, 4, 6, 8, 10]\n"
     ]
    }
   ],
   "source": [
    "# Sorting a list of numbers\n",
    "numlist = [8,10,6,4,0,2]\n",
    "rank = rank_sequences(numlist) # the algorithm also works for any list of camparable objects\n",
    "n=len(numlist)\n",
    "\n",
    "# To sort the numbers, we simple make a boolean sum of them all that only keeps the one with the correct rank\n",
    "sortedlist = [ np.sum([numlist[k]*(i==rank[k]) for k in range(n)]) for i in range(n) ]\n",
    "print('sorted :',sortedlist)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For a lists of elements of length `n`, we need to compute `n*n` steps of `numlist[k]*(i==rank[k])` in addition to the ranking algorithm. But recall we want to sort sequences. We can do the same for sequences by also looping on the index `j` in each sequence. When sequences have different lengths, we will first equalize their lengths by appending empty characters at the end of the shortest ones with `adjust_lenghts`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def adjust_lenghts(seqlist):\n",
    "    \"\"\"\n",
    "    Append empty characters to short sequences\n",
    "    so they all have the same length\n",
    "    \"\"\"\n",
    "    lengths = [ len(s) for s in seqlist ]\n",
    "    maxlen = np.max(lengths)\n",
    "    newlist = []\n",
    "    for s in seqlist:\n",
    "        if len(s) < maxlen:\n",
    "            if isinstance(s, FheSeq) or isinstance(s, FheMutableSeq):\n",
    "                s += FheSeq(fhe.zeros(maxlen-len(s)), s.get_seq_interface())\n",
    "            elif isinstance(s, Seq) or isinstance(s, MutableSeq):\n",
    "                s += Seq('\\0'*(maxlen-len(s)))\n",
    "            else:\n",
    "                return NotImplemented\n",
    "        newlist.append(s)\n",
    "    \n",
    "    return newlist\n",
    "\n",
    "def sort_sequences(seqlist, ascending=True):\n",
    "    \"\"\"\n",
    "    Sort sequences\n",
    "    \"\"\"\n",
    "    # Rank the sequences\n",
    "    rank = rank_sequences(seqlist, ascending)\n",
    "    \n",
    "    # Adjust their length to be equal if needed\n",
    "    seqlist = adjust_lenghts(seqlist)\n",
    "    \n",
    "    # Create a new list to hold the result\n",
    "    n=len(seqlist)\n",
    "    m=len(seqlist[0])\n",
    "    sorted_list = []\n",
    "    \n",
    "    if isinstance(seqlist[0], FheSeq) or isinstance(seqlist[0], FheMutableSeq):\n",
    "        lib = fhe\n",
    "        seq_interface = seqlist[0].get_seq_interface()\n",
    "    else:\n",
    "        lib = np\n",
    "    \n",
    "    # Fill the array with a boolean sum of all sequences at each row\n",
    "    for i in range(n): # i is the index of the sequence in the sorted list\n",
    "        i_eq_rank_k = rank==i\n",
    "        sorted_array = lib.zeros(m) if lib is fhe else ''\n",
    "        for j in range(m): # j is the index of the letter in the sequence\n",
    "            if lib is fhe:\n",
    "                # k is the index of a sequence in the unsorted list\n",
    "                # we could also put seqlist inside a 2d array and use a dot product here\n",
    "                sorted_array[j] = np.sum( [seqlist[k][j]*i_eq_rank_k[k] for k in range(n)] )\n",
    "            else:\n",
    "                for k in range(n):\n",
    "                    sorted_array += seqlist[k][j]*int(i_eq_rank_k[k])\n",
    "        \n",
    "        # Create sequence object from this array\n",
    "        if lib is fhe:\n",
    "            sorted_list.append(seq_interface.FheSeq_from(sorted_array))\n",
    "        else:\n",
    "            sorted_list.append(Seq(sorted_array))\n",
    "\n",
    "    # Return the list of sorted (adjusted) sequences\n",
    "    return sorted_list\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, we want to make again a `process_sequence` that will work both in python and FHE. Because `sort_sequences` returns a list of sequences, we will need to concatenate them again in FHE and return an array so it can be decrypted."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The sorted list of sequences is:\n",
      " [Seq('AA\u0000'), Seq('AAA'), Seq('AAB'), Seq('CAA'), Seq('CAA')]\n"
     ]
    }
   ],
   "source": [
    "# Test the same sequence list\n",
    "SEQ1 = Seq('AA')\n",
    "SEQ2 = Seq('AAA')\n",
    "SEQ3 = Seq('AAB')\n",
    "SEQ4 = Seq('CAA')\n",
    "SEQ5 = Seq('CAA')\n",
    "\n",
    "seqlist = [SEQ5, SEQ3, SEQ4, SEQ1, SEQ2]\n",
    "\n",
    "# Concatenate sequences using the + operator\n",
    "def concat_sequences(seqlist):\n",
    "    concat_seq = seqlist[0]   \n",
    "    for seq in seqlist[1:]:\n",
    "        concat_seq = concat_seq + seq\n",
    "    return concat_seq\n",
    "\n",
    "def process_sequences(*seqlist):\n",
    "    sorted_list = sort_sequences(seqlist)\n",
    "    if isinstance(seqlist[0], FheSeq) or isinstance(seqlist[0], FheMutableSeq):\n",
    "        return concat_sequences(sorted_list).to_array()\n",
    "    elif isinstance(seqlist[0], Seq) or isinstance(seqlist[0], MutableSeq):\n",
    "        return sorted_list\n",
    "    else:\n",
    "        return NotImplemented\n",
    "\n",
    "OUTPUT_PYTHON = process_sequences(*seqlist)\n",
    "print('The sorted list of sequences is:\\n', OUTPUT_PYTHON)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The sorted sequences as array is:\n",
      "|  Compiling  : 3.84 s  |\n",
      "|  Encrypting  : 1.09 s  |\n",
      "|  Running  : 3.12 s  |\n",
      "|  Decrypting  : 0.00 s  |\n",
      "==> Result : [0 0 0 0 0 0 0 0 1 2 0 0 2 0 0] \n",
      "\n",
      "The sorted list of sequences is:\n",
      " [Seq('AAA'), Seq('AAA'), Seq('AAB'), Seq('CAA'), Seq('CAA')]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "KeySetCache: miss, regenerating .keys/14202293608837441708/0_0\n"
     ]
    }
   ],
   "source": [
    "def slice_sequence_n(seq, n):\n",
    "    \"\"\"\n",
    "    Slice a sequence into n subsequences of equal size\n",
    "    \"\"\"\n",
    "    seqs=[]\n",
    "    index=0\n",
    "    length = len(seq)//n\n",
    "    while index < len(seq):\n",
    "        seqs.append(seq[index:index+length])\n",
    "        index+=length\n",
    "    return seqs\n",
    "\n",
    "seq_interface = SeqInterface(\"ABCD\")\n",
    "\n",
    "# test again on a limited alphabet to be faster\n",
    "sorted_seq = compute_fhe_output(seqlist, process_sequences, seq_output = False,\n",
    "             seq_interface = seq_interface,\n",
    "             description='The sorted sequences as array is:', res=True)\n",
    "\n",
    "OUTPUT_FHE = [seq_interface.array_to_Seq(s) for s in slice_sequence_n(sorted_seq, len(seqlist))]\n",
    "print('The sorted list of sequences is:\\n', OUTPUT_FHE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"section-v\"></a>\n",
    "## V. Aligning sequences\n",
    "\n",
    "Sequence alignment algorithms are extensively employed in the field of bioinformatics in order to identify patterns of similarity between sequences. These algorithms come in two main types: local and global alignment. Local alignment algorithms are designed to discover optimal local matches between two sequences, such as pinpointing specific a common segment within longer gene sequences. Conversely, global alignment algorithms strive to find the best alignment across the entire length of sequences, making them suitable for comparing different versions of the same gene, for example.\n",
    "\n",
    "<a id=\"section-v-1\"></a>\n",
    "### Local alignment with the Smith-Waterman algorithm\n",
    "\n",
    "( difficulty level: hard )\n",
    "\n",
    "The [Smith-Waterman](https://en.wikipedia.org/wiki/Smith%E2%80%93Waterman_algorithm) algorithm is a local alignment algorithm. Given two sequences, it will look for two segments in each sequence that have a minimal cost for transforming one into the other, and it will output how to transform one segment into the other and vice-versa. The transformation of a segment into another is made with the following possible steps:\n",
    "- no transformation, the characters are matching\n",
    "- swapping the characters, they are non matching\n",
    "- skipping a character in one segment only, this is called a gap\n",
    "\n",
    "Each possible step is attributed a scoring, for instance:\n",
    "```python\n",
    "class Score:\n",
    "    MATCH = 1\n",
    "    MISMATCH = -1\n",
    "    GAP = -1\n",
    "```\n",
    "\n",
    "To transform the segment `'CAG'` into the segment `'ACGA'` for example, we can skip `C` in the first segment (gap), match the two `A`s (match), skip the `C` in the second segment (gap), match the `G`s and finally skip the `A` in the second segment (gap). This gives a total score of `-1` (three gaps and two matches, no mismatches). The transformation, called an alignment, is written like this (the `-` character being used for gaps):\n",
    "```\n",
    "CA-G-\n",
    "-ACGA\n",
    "``` \n",
    "\n",
    "The idea of the **Smith-Waterman** algorithm is the following (see figure):\n",
    "1. Compute a grid (the **score  matrix**), from top-left to bottom-right cell, containing all the possible local alignment scores between each pair of characters, and keep track of the total maximum score in the grid and its location. At the same time, keep a trace in another grid (the **trace matrix**) of the direction to take in order to reach the next best score between the previously computed grid cells of the score matrix in the **LEFT**, **UP**, and **DIAGONAL** directions (**LEFT = 1**, **UP = 2** and **DIAGONAL = 3**). An additional first row and column are filled with zeros in the score matrix. If the path arrives there in the trace matrix, it will stop.\n",
    "2. Start at the location of the **total maximum score** of the score matrix, in the trace matrix, and follow the backward path indicated by the direction values in the trace matrix. This will draw a path of a local alignment between two segments of the sequences that maximise the alignment score. The alignment sequences are computed backwards. As the algorithm is local, this path may be very short, starting from the maximum score value and stopping whenever the score gets to a stop value **STOP = 0**. More precisions about the mathematical foundations [here](https://en.wikipedia.org/wiki/Smith%E2%80%93Waterman_algorithm#).\n",
    "\n",
    "<div style=\"margin-top: 40px; margin-bottom: 20px;\">\n",
    "<img src=\"https://rcd-media.com/docs/fhe/mat_1.jpg\" width=\"600\"/>\n",
    "</div>\n",
    "<div style=\"text-align: center;  margin-left: 50px; margin-top: 20px; margin-bottom: 45px;\">\n",
    "  <h3>Smith Waterman matrices for sequences 'ATACAGT' and 'GGATGAC'</h3>\n",
    "</div>\n",
    "\n",
    "Let's implement the algorithm in a pure python and FHE compatible way, and test it on the same example as in the figure:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def max_list(l, strictly_gt=True):\n",
    "    \"\"\"\n",
    "    Return the maximum value in a list using boolean mutliplication,\n",
    "    along with the argmax (the index of the maximum)\n",
    "    \n",
    "    strictly_gt: set to True, it will return the index of the first maximum found in case there are\n",
    "    several identical values, while set to False, it will return the index of the last maximum\n",
    "    \"\"\"\n",
    "    maxval = l[0]\n",
    "    argmax = 0\n",
    "    for i in range(1, len(l)):\n",
    "        is_gt = l[i]>maxval if strictly_gt else l[i]>=maxval\n",
    "        maxval = is_gt*l[i] + (1-is_gt)*maxval\n",
    "        argmax = is_gt*i + (1-is_gt)*argmax\n",
    "    \n",
    "    return maxval, argmax\n",
    "\n",
    "# Assigning the constants for the scores\n",
    "# These can be changed depending on what alignment we are looking for\n",
    "class Score:\n",
    "    MATCH = 1\n",
    "    MISMATCH = -1\n",
    "    GAP = -1\n",
    "\n",
    "# Assigning the constant values for the traceback\n",
    "class Trace:\n",
    "    STOP=0\n",
    "    LEFT=1\n",
    "    UP=2\n",
    "    DIAG=3\n",
    "\n",
    "def smith_waterman(seq1, seq2, strictly_gt=True):\n",
    "    \"\"\"\n",
    "    Returns the Smith Waterman local alignment of two sequences\n",
    "    \n",
    "    strictly_gt=True will favor shorter, more exact sequences\n",
    "    while strictly_gt=False will favor longer, less exact sequences\n",
    "    \"\"\"   \n",
    "    ## 1\n",
    "    # Generating the empty matrices for storing scores and tracing\n",
    "    row = len(seq1) + 1\n",
    "    col = len(seq2) + 1\n",
    "    score_matrix = fhe.zeros((row, col))\n",
    "    tracing_matrix = fhe.zeros((row, col))\n",
    "    \n",
    "    # set variables depending on python or FHE run:\n",
    "    if isinstance(seq1, FheSeq) or isinstance(seq1, FheMutableSeq):\n",
    "        seq_interface = seq1.get_seq_interface()\n",
    "        zero_val = fhe.zero()\n",
    "        empty_char =  seq_interface.letter_to_integer('\\0')\n",
    "        gap_char = seq_interface.letter_to_integer('-')\n",
    "        make_empty_seq = lambda x: FheMutableSeq(fhe.zeros(x), seq_interface) # we need mutable sequences \n",
    "    elif isinstance(seq1, Seq) or isinstance(seq1, MutableSeq):\n",
    "        zero_val = 0\n",
    "        empty_char = '\\0'\n",
    "        gap_char = '-'\n",
    "        make_empty_seq = lambda x: MutableSeq('\\0'*x)        \n",
    "    else:\n",
    "        return NotImplemented\n",
    "        \n",
    "    # Initializing the variables to find the highest scoring cell\n",
    "    max_score = max_i = max_j = -1\n",
    "    \n",
    "    # Calculating the scores for all cells in the score matrix\n",
    "    for i in range(1, row):\n",
    "        for j in range(1, col):\n",
    "            # Calculating the diagonal score (match score)\n",
    "            is_eq = seq1[i - 1] == seq2[j - 1]\n",
    "            match_value = Score.MATCH * is_eq + Score.MISMATCH* (1-is_eq)\n",
    "            diagonal_score = score_matrix[i - 1, j - 1] + match_value\n",
    "            \n",
    "            # Calculating the vertical gap score\n",
    "            vertical_score = score_matrix[i - 1, j] + Score.GAP\n",
    "            \n",
    "            # Calculating the horizontal gap score\n",
    "            horizontal_score = score_matrix[i, j - 1] + Score.GAP\n",
    "            \n",
    "            # Taking the local highest score and where it comes from\n",
    "            value_list = [zero_val, horizontal_score, vertical_score, diagonal_score]\n",
    "            score_matrix[i, j], tracing_matrix[i, j] = max_list(value_list, strictly_gt)\n",
    "            # if some scores are identical and are the maximum value, the value tracing_matrix[i,j] will\n",
    "            # depend on the parameter strictly_gt (see max_list function)\n",
    "\n",
    "            # now, tracing_matrix[i, j] = 0, 1, 2 or 3, meaning STOP, LEFT, UP or DIAG\n",
    "                \n",
    "            # Tracking the cell with the global maximum score\n",
    "            is_ge = score_matrix[i, j] >= max_score\n",
    "            not_is_ge = (1-is_ge)\n",
    "            max_i = max_i * not_is_ge + i * is_ge\n",
    "            max_j = max_j * not_is_ge + j * is_ge\n",
    "            max_score = max_score * not_is_ge + score_matrix[i, j] * is_ge       \n",
    "    \n",
    "    if isinstance(seq1, Seq) or isinstance(seq1, MutableSeq):\n",
    "        print('Smith Waterman local alignment')\n",
    "        print('\\nScore matrix :')\n",
    "        print(score_matrix)\n",
    "        print('\\nTracing matrix :')\n",
    "        print(tracing_matrix)\n",
    "        \n",
    "    ## 2\n",
    "    # Initializing the variables for tracing\n",
    "    len_out = max(row,col)-1\n",
    "    aligned_seq1 = make_empty_seq(len_out)\n",
    "    aligned_seq2 = make_empty_seq(len_out)\n",
    "    ind_ij = fhe.zeros((row, col))\n",
    "    \n",
    "    # Tracing and computing the pathway with the local alignment from the global maximum\n",
    "    #  while tracing_matrix[max_i, max_j] != Trace.STOP:\n",
    "    for k in reversed(range(len_out)):\n",
    "\n",
    "        # select value in tracing_matrix at encrypted indices max_i, max_j:\n",
    "        ind_i = np.arange(row)==max_i\n",
    "        ind_j = np.arange(col)==max_j\n",
    "        \n",
    "        for i in range(0, row):\n",
    "            for j in range(0, col):\n",
    "                ind_ij[i,j] = ind_i[i] & ind_j[j] # using & for binary is faster than *\n",
    "        \n",
    "        V = np.sum(tracing_matrix*ind_ij)\n",
    "\n",
    "        VSTOP = (V==Trace.STOP)*1\n",
    "        VLEFT = (V==Trace.LEFT)*1\n",
    "        VUP = (V==Trace.UP)*1\n",
    "        VDIAG = (V==Trace.DIAG)*1\n",
    "        VLEFT_or_VDIAG = VLEFT + VDIAG\n",
    "        VUP_or_VDIAG = VUP + VDIAG        \n",
    "        # Note: the *1 is just for pure python, to convert booleans into ints\n",
    "        \n",
    "        # select value in seq1 at encrypted index max_i - 1:\n",
    "        v1 = seq1[max_i-1]\n",
    "        # Note: we accessed seq1 with an encrypted index max_i-1, this is allowed by the FheSeq class\n",
    "        # but it is computationally expensive, see FheSeq.__getitem__ function \n",
    "                \n",
    "        current_aligned_char1 = (VUP_or_VDIAG)*v1 + VLEFT*gap_char + VSTOP*empty_char;\n",
    "        # Note: in FHE, one should make as few multiplications as possible for speed purpose,\n",
    "        # that's why we regrouped VDIAG + VUP here instead of VDIAG*v1 + VUP*v1\n",
    "        \n",
    "        # select value in seq2 at encrypted index max_j - 1:\n",
    "        v2 = seq2[max_j-1]\n",
    "        current_aligned_char2 = (VLEFT_or_VDIAG)*v2 + VUP*gap_char + VSTOP*empty_char; \n",
    "        \n",
    "        max_i = max_i - (VUP_or_VDIAG)\n",
    "        max_j = max_j - (VLEFT_or_VDIAG)\n",
    "        \n",
    "        aligned_seq1[k] = current_aligned_char1\n",
    "        aligned_seq2[k] = current_aligned_char2\n",
    "\n",
    "    return aligned_seq1, aligned_seq2\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In pure python, a stop condition can be verified and lead to the effective end of the algorithm, producing alignment sequences of variable sizes. However, as you can see, in FHE a condition cannot lead to the end of the algorithm. Thus, instead of stopping the path, a stop value will make the path stagnate, and it will append empty characters to the aligned sequences. This will lead to fixed-size sequences with apparent variable sizes when printed.\n",
    "\n",
    "Also note that it is possible to access FheSeq values at encrypted index (see `v1 = seq1[max_i-1]`). It uses a trick, the same we used to select the value in the trace matrix at encrypted indices `max_i, max_j`. This trick is very slow to process compared to unencrypted indexing. That's why the complexity of this algorithm is normally in **O(mn)**, but in FHE, because of the loops for the trick, it is actually in **O(p^3)** where `p=max(m,n)`.\n",
    "\n",
    "Let's first test the algorithm in pure python, on the same sequences as in the previous figure. The printed matrices should be identical:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Smith Waterman local alignment\n",
      "\n",
      "Score matrix :\n",
      "[[0 0 0 0 0 0 0 0]\n",
      " [0 0 0 1 0 0 1 0]\n",
      " [0 0 0 0 2 1 0 0]\n",
      " [0 0 0 1 1 1 2 1]\n",
      " [0 0 0 0 0 0 1 3]\n",
      " [0 0 0 1 0 0 1 2]\n",
      " [0 1 1 0 0 1 0 1]\n",
      " [0 0 0 0 1 0 0 0]]\n",
      "\n",
      "Tracing matrix :\n",
      "[[0 0 0 0 0 0 0 0]\n",
      " [0 0 0 3 0 0 3 0]\n",
      " [0 0 0 0 3 1 0 0]\n",
      " [0 0 0 3 2 3 3 1]\n",
      " [0 0 0 0 0 0 2 3]\n",
      " [0 0 0 3 0 0 3 2]\n",
      " [0 3 3 0 0 3 0 2]\n",
      " [0 0 0 0 3 0 0 0]]\n",
      "\n",
      "The aligned sequences are:\n",
      " \u0000\u0000AT-AC \n",
      " \u0000\u0000ATGAC\n"
     ]
    }
   ],
   "source": [
    "SEQ1 = Seq('ATACAGT')\n",
    "SEQ2 = Seq('GGATGAC')\n",
    "\n",
    "seqlist = [SEQ1, SEQ2]\n",
    "\n",
    "def process_sequence(*seqlist):\n",
    "    if isinstance(seqlist[0], FheSeq) or isinstance(seqlist[0], FheMutableSeq):\n",
    "        aligned_seq1, aligned_seq2 = smith_waterman(seqlist[0],seqlist[1])\n",
    "        newseq = concat_sequences([aligned_seq1, aligned_seq2])\n",
    "        return newseq.to_array()\n",
    "    elif isinstance(seqlist[0], Seq) or isinstance(seqlist[0], MutableSeq):\n",
    "        aligned_seq1, aligned_seq2 = smith_waterman(seqlist[0], seqlist[1])\n",
    "        return [Seq(aligned_seq1), Seq(aligned_seq2)]\n",
    "    else:\n",
    "        return NotImplemented\n",
    "    \n",
    "OUTPUT_PYTHON = process_sequence(*seqlist)\n",
    "print('\\nThe aligned sequences are:\\n', OUTPUT_PYTHON[0], '\\n', OUTPUT_PYTHON[1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remember, the actual output aligned sequences have a fixed length, but they may start with the empty character `'\\0'` repeated several times, which does not show when printed.  \n",
    "\n",
    "We can now test it in FHE, starting with shorter sequences for a faster computation (remember the very high algorithmic complexity of the algorithm): "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FHE Smith-Waterman\n",
      "|  Compiling  : 7.33 s  |\n",
      "|  Encrypting  : 0.17 s  |\n",
      "|  Running  : 7.47 s  |\n",
      "|  Decrypting  : 0.00 s  |\n",
      "==> Result : [0 2 5 0 2 5] \n",
      "\n",
      "The aligned sequences are:\n",
      " \u0000AT \n",
      " \u0000AT\n"
     ]
    }
   ],
   "source": [
    "# we can set shorter sequences for faster computation, because the algorithm is very greedy\n",
    "SEQ1 = Seq('ATG')\n",
    "SEQ2 = Seq('GAT')\n",
    "\n",
    "#SEQ1 = Seq('ATACAGT')\n",
    "#SEQ2 = Seq('GGATGAC')\n",
    "\n",
    "seqlist = [SEQ1, SEQ2]\n",
    "seq_interface = SeqInterface('ACGTU-') # we need the - character as well here\n",
    "aligned_seqs = compute_fhe_output(seqlist, process_sequence, seq_output = False,\n",
    "                                seq_interface = seq_interface,\n",
    "                                description='FHE Smith-Waterman',res=True)\n",
    "len_out = len(aligned_seqs)//2\n",
    "\n",
    "aligned_seq1 = seq_interface.array_to_Seq(aligned_seqs[0:len_out])\n",
    "aligned_seq2 = seq_interface.array_to_Seq(aligned_seqs[len_out:])\n",
    "print('The aligned sequences are:\\n', aligned_seq1, '\\n', aligned_seq2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Going further\n",
    "One can also try different combinations of matching, substitution and gap penalties/scores.\n",
    "\n",
    "For instance, another scoring system that penalizes more gaps might be:\n",
    "\n",
    "```python\n",
    "class Score:\n",
    "    MATCH = 1\n",
    "    MISMATCH = -1\n",
    "    GAP = -2\n",
    "```\n",
    "\n",
    "And another penalizing more mismatches:\n",
    "\n",
    "```python\n",
    "class Score:\n",
    "    MATCH = 1\n",
    "    MISMATCH = -2\n",
    "    GAP = -1\n",
    "```\n",
    "\n",
    "More complicated scoring systems may attribute values not only for the type of transformation, but also for the characters that are involved. For example, a match between `A` and `A` may be given a score of `1`, but a match between `T` and `T` may be given score `4`. The same applies for mismatches and gaps."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"section-v-2\"></a>\n",
    "### Global alignment with the Needleman-Wunsch algorithm\n",
    "\n",
    "( difficulty level: hard )\n",
    "\n",
    "The [Needleman-Wunsch](https://en.wikipedia.org/wiki/Needleman%E2%80%93Wunsch_algorithm) algorithm is a global and optimal alignment algorithm. It is almost identical to the **Smith-Waterman** algorithm. Instead of going from the maximum score cell to a stop value, which leads to short and local alignment sequences, it runs from the bottoms-left cell of the trace matrix up to its top-left cell, leading to long and global alignment sequences. The output of the algorithm are optimal transformations of one whole sequence into the other and vice-versa.\n",
    "\n",
    "Being optimal means that the results are not approximations or best guesses, and that's why the algorithm is widely used for global alignment, but it is also expensive to run for long sequences.\n",
    "\n",
    "<div style=\"margin-top: 30px;\">\n",
    "</div>\n",
    "\n",
    "The two steps for this algorithm are the following:\n",
    "1. The first step is almost identical to the one from **Smith-Waterman**. The first difference is the scoring which allows negative values. Then, the trace matrix has only one stop value on the top-left cell. Finally, at initialization, the first row and column of the two matrices are non zero and set so as to lead the path toward the top-left cell rather than stopping.\n",
    "2. Regarding the second step, let's rewrite and highlight its differences with the one from **Smith-Waterman** that we saw in the previous section: Start at the **bottom-right cell**, in the trace matrix, and follow the backward path indicated by the direction values in the trace matrix. This will draw a path of a **global** alignment between the whole sequences that maximise the alignment score. The alignment sequences are computed backwards. As the algorithm is **global**, this path may be quite long, starting from bottom-right and stopping at top-left. Its length of non empty characters can be between `max(len1, len2)` if the path is fully diagonal, and `len(1) + len(2)` if the path has the shape of a stair. \n",
    "\n",
    "\n",
    "<div style=\"margin-top: 40px; margin-bottom: 20px;\">\n",
    "<img src=\"https://rcd-media.com/docs/fhe/mat_2.jpg\" width=\"600\"/>\n",
    "</div>\n",
    "<div style=\"text-align: center;  margin-left: 50px; margin-top: 20px; margin-bottom: 45px;\">\n",
    "  <h3>Needleman Wunsch matrices for sequences 'ATACAGT' and 'GGATGAC'</h3>\n",
    "</div>\n",
    "\n",
    "Let's implement this new algorithm also in a pure python and FHE compatible way, and test it on the same example as previously. We mark the differences with the previous algorithm in the code with `### DIFF` to make them easy to see:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assigning the constants for the scores\n",
    "# These can be changed depending on what alignment we are looking for\n",
    "class Score:\n",
    "    MATCH = 1\n",
    "    MISMATCH = -1\n",
    "    GAP = -1\n",
    "\n",
    "# Assigning the constant values for the traceback\n",
    "class Trace:\n",
    "    STOP = 0 # STOP is only needed for the FHE version of the algorithm which cannot use a while loop\n",
    "    LEFT = 1 \n",
    "    UP = 2\n",
    "    DIAG = 3\n",
    "\n",
    "def needleman_wunsch(seq1, seq2):\n",
    "    \"\"\"\n",
    "    Returns the Needleman Wunsch global alignment of two sequences\n",
    "    \"\"\"  \n",
    "    ## 1\n",
    "    # Generating the empty matrices for storing scores and tracing\n",
    "    row = len(seq1) + 1\n",
    "    col = len(seq2) + 1\n",
    "    score_matrix = fhe.zeros((row, col))\n",
    "    tracing_matrix = fhe.zeros((row, col))\n",
    "    \n",
    "    # initialize the first row and col of the matrix\n",
    "    ### DIFF (as the algorithm is global, if it reaches the first row or column\n",
    "    #   in the traceback process, it won't stop but rather continue straight to top-left)\n",
    "    score_matrix[:,0] = -1*np.arange(row)\n",
    "    tracing_matrix[1:,0] = 2 # go up if first left column is reached\n",
    "    score_matrix[0,:] = -1*np.arange(col)\n",
    "    tracing_matrix[0,1:] = 1 # go left if first row is reached    \n",
    "    \n",
    "    # set variables depending on python or FHE run:\n",
    "    if isinstance(seq1, FheSeq) or isinstance(seq1, FheMutableSeq):\n",
    "        #   sequence is as long as the longest of the two input sequences)\n",
    "        seq_interface = seq1.get_seq_interface()\n",
    "        empty_char = seq_interface.letter_to_integer('\\0')\n",
    "        gap_char = seq_interface.letter_to_integer('-')\n",
    "        make_empty_seq = lambda x: FheMutableSeq(fhe.zeros(x), seq_interface)\n",
    "    elif isinstance(seq1, Seq) or isinstance(seq1, MutableSeq):\n",
    "        empty_char = '\\0'\n",
    "        gap_char = '-'\n",
    "        make_empty_seq = lambda x: MutableSeq('\\0'*x)\n",
    "    else:\n",
    "        return NotImplemented\n",
    "        \n",
    "    # Initializing the variables to find the highest scoring cell\n",
    "    max_score = max_i = max_j = -1\n",
    "    \n",
    "    # Calculating the scores for all cells in the matrix\n",
    "    for i in range(1, row):\n",
    "        for j in range(1, col):\n",
    "            # Calculating the diagonal score (match score)\n",
    "            is_eq = seq1[i - 1] == seq2[j - 1]\n",
    "            match_value = Score.MATCH * is_eq + Score.MISMATCH* (1-is_eq)\n",
    "            diagonal_score = score_matrix[i - 1, j - 1] + match_value\n",
    "            \n",
    "            # Calculating the vertical gap score\n",
    "            vertical_score = score_matrix[i - 1, j] + Score.GAP\n",
    "            \n",
    "            # Calculating the horizontal gap score\n",
    "            horizontal_score = score_matrix[i, j - 1] + Score.GAP\n",
    "            \n",
    "            # Taking the local highest score and where it comes from\n",
    "            values = fhe.array([horizontal_score, vertical_score, diagonal_score])\n",
    "            ### DIFF (the value list has only 3 elements here, without a stopping option)\n",
    "            score_matrix[i, j], tracing_matrix[i, j] = max_list(values, strictly_gt=True)\n",
    "            ### DIFF (strictly_gt=False ensures the diagonal direction will be favored\n",
    "            #   in case the score is the same as vertical or horizontal, see max_list function)    \n",
    "            tracing_matrix[i, j] += 1 # the argmax of the list is 0,1,2 and we want 1,2,3\n",
    "            \n",
    "            # now, tracing_matrix[i, j] = 0, 1, 2 or 3, meaning STOP, LEFT, UP or DIAG\n",
    "                \n",
    "            ### DIFF (No tracking of the cell with the global maximum score here)\n",
    " \n",
    "    if isinstance(seq1, Seq) or isinstance(seq1, MutableSeq):\n",
    "        print('Needleman-Wunsch global alignment')\n",
    "        print('\\nScore matrix :')\n",
    "        print(score_matrix)\n",
    "        print('\\ntracing matrix :')\n",
    "        print(tracing_matrix)\n",
    "    \n",
    "    ## 2\n",
    "    # Initializing the variables for tracing\n",
    "    ### DIFF (the maximum possible length is the sum of the lengths when the path has a stair shape)\n",
    "    len_out = len(seq1)+len(seq2)\n",
    "    aligned_seq1 = make_empty_seq(len_out)\n",
    "    aligned_seq2 = make_empty_seq(len_out)\n",
    "    ind_ij = fhe.zeros((row, col))\n",
    "    \n",
    "    # Tracing and computing the pathway with the local alignment from the bottom-left up\n",
    "    ### DIFF (the traceback process starts at the bottom-right corner of the matrix)    \n",
    "    max_i = row-1\n",
    "    max_j = col-1\n",
    "    \n",
    "    # select the starting value in tracing_matrix at index max_i, max_j (unencrypted for now)\n",
    "    V = tracing_matrix[max_i, max_j]\n",
    "        \n",
    "    for k in reversed(range(len_out)):\n",
    "  \n",
    "        VSTOP = (V==Trace.STOP)*1 # this will happen only when max_i = max_j = 0        \n",
    "        VLEFT = (V==Trace.LEFT)*1\n",
    "        VUP = (V==Trace.UP)*1\n",
    "        VDIAG = (V==Trace.DIAG)*1\n",
    "        VLEFT_or_VDIAG = VLEFT + VDIAG\n",
    "        VUP_or_VDIAG = VUP + VDIAG\n",
    "        # Note: the *1 is just for pure python, to convert booleans into ints\n",
    "                        \n",
    "        # select value in seq1 at encrypted index max_i - 1:\n",
    "        v1 = seq1[max_i-1]\n",
    "        # Note: we accessed seq1 with an encrypted index max_i-1, this is allowed by the FheSeq class\n",
    "        # but it is computationally expensive, see FheSeq.__getitem__ function                 \n",
    "        current_aligned_char1 = (VUP_or_VDIAG)*v1 + VLEFT*gap_char + VSTOP*empty_char;\n",
    "        # Note: in FHE, one should make as few multiplications as possible for speed purpose,\n",
    "        # that's why we regrouped (VDIAG + VUP)*v1 here, instead of VDIAG*v1 + VUP*v1\n",
    "        \n",
    "        # select value in seq2 at encrypted index max_j - 1:\n",
    "        v2 = seq2[max_j-1]\n",
    "        current_aligned_char2 = (VLEFT_or_VDIAG)*v2 + VUP*gap_char + VSTOP*empty_char; \n",
    "        \n",
    "        max_i = max_i - (VUP_or_VDIAG)\n",
    "        max_j = max_j - (VLEFT_or_VDIAG)\n",
    "        \n",
    "        aligned_seq1[k] = current_aligned_char1\n",
    "        aligned_seq2[k] = current_aligned_char2\n",
    "        \n",
    "        # select value in tracing_matrix at encrypted indices max_i, max_j:\n",
    "        ind_i = (np.arange(row)==max_i)*1\n",
    "        ind_j = (np.arange(col)==max_j)*1\n",
    "        \n",
    "        if k == 0: # no need to compute the last one\n",
    "            break\n",
    "            \n",
    "        ### DIFF (as the first indices where unencrypted, we prevented 1 costly computation)\n",
    "        for i in range(0, row):\n",
    "            for j in range(0, col):\n",
    "                ind_ij[i,j] = ind_i[i] & ind_j[j]\n",
    "        \n",
    "        V = np.sum(tracing_matrix*ind_ij)  \n",
    "        \n",
    "    return aligned_seq1, aligned_seq2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's now also first test the algorithm in pure python, on the same sequences as in the previous figure. The printed matrices should be identical to the figure of this section. We will also compute a `similarity_score` before and after alignment, to verify that the algorithm actually increases characters alignment:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Needleman-Wunsch global alignment\n",
      "\n",
      "Score matrix :\n",
      "[[ 0 -1 -2 -3 -4 -5 -6 -7]\n",
      " [-1 -1 -2 -1 -2 -3 -4 -5]\n",
      " [-2 -2 -2 -2  0 -1 -2 -3]\n",
      " [-3 -3 -3 -1 -1 -1  0 -1]\n",
      " [-4 -4 -4 -2 -2 -2 -1  1]\n",
      " [-5 -5 -5 -3 -3 -3 -1  0]\n",
      " [-6 -4 -4 -4 -4 -2 -2 -1]\n",
      " [-7 -5 -5 -5 -3 -3 -3 -2]]\n",
      "\n",
      "tracing matrix :\n",
      "[[0 1 1 1 1 1 1 1]\n",
      " [2 3 1 3 1 1 1 1]\n",
      " [2 2 3 2 3 1 1 1]\n",
      " [2 2 2 3 2 3 3 1]\n",
      " [2 2 2 2 2 2 2 3]\n",
      " [2 2 2 2 2 2 3 2]\n",
      " [2 3 3 2 2 3 2 2]\n",
      " [2 2 2 2 3 2 2 2]]\n",
      "\n",
      "The aligned sequences are:\n",
      " \u0000\u0000\u0000\u0000--AT-ACAGT \n",
      " \u0000\u0000\u0000\u0000GGATGAC---\n",
      "\n",
      "Similarity score before alignment : 1\n",
      "Similarity score after alignment : 4\n"
     ]
    }
   ],
   "source": [
    "def similarity_score(seq1, seq2):\n",
    "    \"\"\"\n",
    "    Compute the number of identical characters in two aligned sequences\n",
    "    \"\"\"\n",
    "    nequal = 0\n",
    "    for i in range(min(len(seq1), len(seq2))):\n",
    "        nequal += 1 if seq1[i]==seq2[i] and seq1[i]!='-' and seq1[i]!='\\0' else 0\n",
    "    return nequal\n",
    "\n",
    "SEQ1 = Seq('ATACAGT')\n",
    "SEQ2 = Seq('GGATGAC')\n",
    "\n",
    "seqlist = [SEQ1, SEQ2]\n",
    "\n",
    "def process_sequence(*seqlist):\n",
    "    if isinstance(seqlist[0], FheSeq) or isinstance(seqlist[0], FheMutableSeq):\n",
    "        aligned_seq1, aligned_seq2 = needleman_wunsch(seqlist[0],seqlist[1])\n",
    "        newseq = concat_sequences([aligned_seq1, aligned_seq2])\n",
    "        return newseq.to_array()\n",
    "    elif isinstance(seqlist[0], Seq) or isinstance(seq_liseqlistst[0], MutableSeq):\n",
    "        aligned_seq1, aligned_seq2 = needleman_wunsch(seqlist[0], seqlist[1])\n",
    "        return [Seq(aligned_seq1), Seq(aligned_seq2)]\n",
    "    else:\n",
    "        return NotImplemented\n",
    "    \n",
    "OUTPUT_PYTHON = process_sequence(*seqlist)\n",
    "print('\\nThe aligned sequences are:\\n', OUTPUT_PYTHON[0], '\\n', OUTPUT_PYTHON[1])\n",
    "print('\\nSimilarity score before alignment :', similarity_score(SEQ1, SEQ2) )\n",
    "print('Similarity score after alignment :', similarity_score(*OUTPUT_PYTHON))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remember again, the actual output aligned sequences have also a fixed length here, but they may start with the empty character `'\\0'` repeated several times, which does not show when printed.  \n",
    "\n",
    "We can now test it again in FHE, starting as well with shorter sequences for a faster computation:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Needleman-Wunsch :\n",
      "|  Compiling  : 13.71 s  |\n",
      "|  Encrypting  : 0.28 s  |\n",
      "|  Running  : 6.61 s  |\n",
      "|  Decrypting  : 0.00 s  |\n",
      "==> Result : [0 0 1 2 5 4 0 0 4 2 5 1] \n",
      "\n",
      "\n",
      "The aligned sequences are:\n",
      " \u0000\u0000-ATG \n",
      " \u0000\u0000GAT-\n",
      "\n",
      "Similarity score before alignment : 0\n",
      "Similarity score after alignment : 2\n"
     ]
    }
   ],
   "source": [
    "# we can set shorter sequences for faster computation, because the algorithm is very greedy\n",
    "SEQ1 = Seq('ATG')\n",
    "SEQ2 = Seq('GAT')\n",
    "\n",
    "#SEQ1 = Seq('ATTACAG')\n",
    "#SEQ2 = Seq('GGATGAC')\n",
    "\n",
    "seq_list = [SEQ1, SEQ2]\n",
    "\n",
    "seq_interface = SeqInterface('ACGT-')\n",
    "aligned_seqs = compute_fhe_output(seq_list, process_sequence, seq_output = False,\n",
    "                                  seq_interface=seq_interface,\n",
    "                                  description='Needleman-Wunsch :',res=True)\n",
    "len_out = len(aligned_seqs)//2\n",
    "\n",
    "aligned_seq1 = seq_interface.array_to_Seq(aligned_seqs[0:len_out])\n",
    "aligned_seq2 = seq_interface.array_to_Seq(aligned_seqs[len_out:])\n",
    "\n",
    "print('\\nThe aligned sequences are:\\n', aligned_seq1,'\\n', aligned_seq2)\n",
    "print('\\nSimilarity score before alignment :', similarity_score(SEQ1, SEQ2) )\n",
    "print('Similarity score after alignment :', similarity_score(aligned_seq1, aligned_seq2))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, one can also try different combinations of matching, substitution and gap penalties/scores to get different behaviors of the algorithm."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"section-v-3\"></a>\n",
    "### Going further: alignment software used in the industry\n",
    "\n",
    "The two algorithm we saw have both positive and negative aspects:\n",
    "- The **Needleman-Wunsch** algorithm is quite greedy, particularly in **FHE**, so it takes time to compute global alignment on long sequences. However, it is optimal and generally used on relatively short sequences.\n",
    "- The **Smith-Waterman** algorithm is as greedy as the **Needleman-Wunsch**, and will only output a single aligned sequence. It is a major problem in bioinformatics where we usually want to retrieve as much local alignments as possible within a given score threshold. Also, while the first sequence (the **query** sequence) is generally quite small, we want to look for it into a potentially very long sequence (the **reference**).\n",
    "\n",
    "Numeral [alignment softwares](https://en.wikipedia.org/wiki/List_of_sequence_alignment_software#Short-read_sequence_alignment) have been designed to serve diverse objectives. Some of the most famous and widely used are **BLAST** (Basic Local Alignment Search Too) and **BWA** (Burrowsâ€“Wheeler aligner), which are **local alignment** algorithm and both come with many variations. They enable rapid searching for a query sequence within a considerably long reference sequence, sometimes even encompassing entire databases of sequences. They retrieve several alignement sequences which scores are over a given threshold.\n",
    "\n",
    "Their main principle is similar: first run once a heavy computation so as to prepare the **reference** sequence for optimal search, then look quickly for any **queries**, with or without exact match, and using statistical heuristics. The preparation step takes a lot of time, but then, it allows to run as many quick queries as needed. The human genome is for instance a common reference that researchers use to look for gene variants queries.\n",
    "\n",
    "In **2023**, none of these complexe algorithms exist yet in a **FHE** version. A number of important steps must be taken first, beginning with a hardware acceleration allowing to run them in a decent time and with long reference sequences. A FHE compatible data structure similar to the python dictionary is also required."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"section-v-4\"></a>\n",
    "### Bonus: improving the speed of the Levenshtein distance\n",
    "\n",
    "The **Needleman-Wunsch** algorithm is related to the **Levenshtein distance** we saw earlier. Remember, the **Levenshtein** distance measured the number of operations required to transform a sequence into another one. The difference relies in the type of operations that can be used: in the **Levenshtein** algorithm, characters can be deleted while they can only be skipped here. So, though the results are not perfectly related, a small **Levenshtein distance** between two sequences will be associated with a good **Needleman-Wunsch** alignment score of these two sequences.\n",
    "\n",
    "But if they are related, why are the algorithms so different, you might ask? Well, we previously saw the naive, recursive version of the **Levenshtein** distance algorithm. But now that we understood how the **Needleman-Wunsch** worked, we are able to redesign the **Levenshtein** distance algorithm in a non recursive but iterative fashion, which will greatly improve its speed (see [wikipedia](https://en.wikipedia.org/wiki/Levenshtein_distance))."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Levenshtein iterative algorithm\n",
      "\n",
      "Distance matrix :\n",
      "[[0 1 2 3 4]\n",
      " [1 1 2 3 3]\n",
      " [2 1 2 3 4]\n",
      " [3 2 1 2 3]\n",
      " [4 3 2 1 2]]\n",
      "\n",
      "\n",
      "Python: Levenshtein distance of S & T: 2\n"
     ]
    }
   ],
   "source": [
    "def iterative_levenshtein(seq1, seq2):\n",
    "    \"\"\"\n",
    "    Returns the Levenstein distance between two sequences\n",
    "    \"\"\"  \n",
    "    # Generating the empty matrices for storing distance\n",
    "    row = len(seq1) + 1\n",
    "    col = len(seq2) + 1\n",
    "    distance_matrix = fhe.zeros((row, col))\n",
    "    \n",
    "    # initialize the first row and col of the matrix\n",
    "    distance_matrix[:,0] = 1*np.arange(row)\n",
    "    distance_matrix[0,:] = 1*np.arange(col)\n",
    "        \n",
    "    # Calculating the scores for all cells in the matrix\n",
    "    for i in range(1, row):\n",
    "        for j in range(1, col):\n",
    "            # Calculating the diagonal score (match score)\n",
    "            substitution_cost = 1*(seq1[i-1] != seq2[j-1])\n",
    "            \n",
    "            distance_matrix[i, j] = min_array(fhe.array([\n",
    "                                        distance_matrix[i-1, j] + 1,                     # deletion\n",
    "                                        distance_matrix[i, j-1] + 1,                     # insertion\n",
    "                                        distance_matrix[i-1, j-1] + substitution_cost])) # substitution\n",
    "            \n",
    "    if isinstance(seq1, Seq) or isinstance(seq1, MutableSeq):\n",
    "        print('Levenshtein iterative algorithm')\n",
    "        print('\\nDistance matrix :')\n",
    "        print(distance_matrix)\n",
    "        print('\\n')\n",
    "        \n",
    "    # return the bottom-right cell value, which is the distance between the two full sequences\n",
    "    return distance_matrix[-1, -1]\n",
    "\n",
    "S = Seq(\"ACGT\")\n",
    "T = Seq(\"CGTA\")\n",
    "\n",
    "seqlist = [S,T]\n",
    "\n",
    "def process_sequence(*seqlist):\n",
    "    return iterative_levenshtein(seqlist[0],seqlist[1])\n",
    "\n",
    "OUTPUT_PYTHON = process_sequence(*seqlist)\n",
    "print('Python: Levenshtein distance of S & T:', OUTPUT_PYTHON)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's try on the same example as earlier, in FHE:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FHE: Levenshtein distance of S & T:\n",
      "|  Compiling  : 3.36 s  |\n",
      "|  Encrypting  : 0.09 s  |\n",
      "|  Running  : 3.85 s  |\n",
      "|  Decrypting  : 0.00 s  |\n",
      "==> Result : 2 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Let's run the computation again (the seq_interface.max_integer will be changed in circuit_factory ):\n",
    "compute_fhe_output(seqlist, process_sequence, seq_output = False,\n",
    "                   seq_interface = SeqInterface(Alphabets.DNA_RNA),\n",
    "                   description='FHE: Levenshtein distance of S & T:')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As expected, the algorithm is way faster, most of all for the compilation !"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "concrete-biopython",
   "language": "python",
   "name": "concrete-biopython"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
